{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Experiementing with sequence filtering**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-02T03:40:32.129098Z",
     "start_time": "2022-08-02T03:40:30.629837Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import time\n",
    "import pandas as pd\n",
    "import pybloom_live as bloom\n",
    "from Bio import SeqIO as seq\n",
    "from Bio.Seq import Seq as Seqbuild\n",
    "import os\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-29T23:52:25.133421Z",
     "start_time": "2022-06-29T23:52:25.067896Z"
    }
   },
   "outputs": [],
   "source": [
    "fasta = \"C:/Users/jonah/School/ASI/bloom/GCF_002973605.1_ASM297360v1_genomic.fna\"\n",
    "# with open(fasta) as handle:\n",
    "#     for record in list(seq.parse(handle,\"fasta\")):\n",
    "#         print(record.id)\n",
    "\n",
    "\n",
    "        \n",
    "with open(fasta) as handle:\n",
    "    records = list(seq.parse(handle,\"fasta\"))\n",
    "        \n",
    "tryme = records[:2]\n",
    "\n",
    "\n",
    "with open(\"testy.fna\", \"w\") as handle:\n",
    "     seq.write(tryme, handle, \"fasta\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-07-01T21:41:21.143152Z",
     "start_time": "2022-07-01T21:41:21.102808Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3606331\n",
      "58\n",
      "69\n"
     ]
    }
   ],
   "source": [
    "bases=0\n",
    "exons=0\n",
    "with open(fasta) as genome_:\n",
    "    for exon in seq.parse(genome_,\"fasta\"):\n",
    "        bases+=len(exon.seq)\n",
    "        exons+=1\n",
    "\n",
    "print(bases)\n",
    "print(exons)\n",
    "\n",
    "import sys\n",
    "print(sys.getsizeof(\"1\"*20))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-05-23T18:37:07.596754Z",
     "start_time": "2022-05-23T18:37:07.560822Z"
    }
   },
   "source": [
    "# to do: \n",
    "\n",
    "create.py file with bloom filter implemented \n",
    "    -eliminate check_read function call at every iteration\n",
    "    -implement one hit pass for read\n",
    "    -\n",
    "\n",
    "implement input for read generators such that the number created is flexible\n",
    "\n",
    "how would I change this to simulate aligning to a reference transcriptome? no need for splice awareness?\n",
    "\n",
    "Using hash set is now the implementation to beat: perhaps creating compressed partial keys can improve hash time, but maybe not...this may also be overkill for bloom filter but could potentially be useful for storing huge numbers of long sequences as we do this with fewer bytes. \n",
    "\n",
    "things to test for sake of understanding runtimes:\n",
    "    time to check set by key length\n",
    "    time to check set by number of partials\n",
    "    check how much one hit stop reduces runtime\n",
    "    why certain combos of indices give us 50% accuracy and how to avoid this\n",
    "    generally better understand best combos to try in order to make sure we don't systematically miss filter keys\n",
    "    should be a function of partial size, read size, genome size, and check indices\n",
    "    whether genome sampling can be used to reduce runtime while still attaining reasonable accuracy\n",
    "    bloom filter effects on accuracy\n",
    "    stacked bloom filter effects on accuracy\n",
    "    whether we have enough memory to store a huge set filter...have only tested up to 30mm\n",
    "    how much java implementation reduces runtime\n",
    "   \n",
    "\n",
    "these only have potential as alternative to stacked filter\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-02T05:35:07.443065Z",
     "start_time": "2022-08-02T05:35:07.409065Z"
    }
   },
   "outputs": [],
   "source": [
    "def create_genome_exons2(total_size, exon_size):\n",
    "    \"\"\"\n",
    "    create syntheic exon to map to\n",
    "    \n",
    "    input:total_size - length of genome\n",
    "    exon_size- length of each exon\n",
    "    output: matrix of exons -string row format\n",
    "    \"\"\"\n",
    "    \n",
    "    exome = np.random.randint(1,5,total_size)\n",
    "    exome = exome.reshape(-1,exon_size)\n",
    "    output = np.empty(exome.shape[0],dtype=\"object\")\n",
    "    \n",
    "    for i in range(exome.shape[0]):\n",
    "        row = exome[i]\n",
    "        \n",
    "        out_string = str()\n",
    "        for j in range(len(row)):\n",
    "            out_string+=(str(row[j]))\n",
    "            \n",
    "        output[i]=out_string\n",
    "    \n",
    "    return output\n",
    "\n",
    "def create_transcriptome2(read_length, reads):\n",
    "    \"\"\"\n",
    "    create synthetic transcriptome\n",
    "    input-read length\n",
    "    input- reads\n",
    "    \n",
    "    output- reads number of random transcripts of read_length\n",
    "    this function is UNNECCESARY\n",
    "    \"\"\"\n",
    "    \n",
    "    transcriptome = np.random.randint(1,5,read_length*reads)\n",
    "    transcriptome = transcriptome.reshape(-1,reads)\n",
    "    output=np.empty(transcriptome.shape[0],dtype=\"object\")\n",
    "    \n",
    "    for i in range(int(transcriptome.shape[0]/10)):\n",
    "        \n",
    "        row=transcriptome[i]\n",
    "        outstring=str()\n",
    "        \n",
    "        for j in range(len(row)):\n",
    "            outstring+=str(row[j])\n",
    "            \n",
    "        output[i]=outstring\n",
    "            \n",
    "    return output\n",
    "\n",
    "def create_matched_transcriptome2(exome, read_length):\n",
    "    \"\"\"\n",
    "    create synthetic matched reads\n",
    "    \n",
    "    input: exome (mxn) where n is read string length and m is number of array entries\n",
    "    input: read_length: int\n",
    "    \n",
    "    output: matched reads potentially spanning exomes\n",
    "    \"\"\"\n",
    "    \n",
    "    bigstring = str()\n",
    "    \n",
    "    for i in exome[:int(len(exome))]:\n",
    "        bigstring+=i\n",
    "        \n",
    "    output = np.empty(len(bigstring)-read_length+1, dtype = \"object\")\n",
    "    \n",
    "    for i in range(len(bigstring)-read_length+1):\n",
    "        transcript = bigstring[i:i+read_length]\n",
    "        output[i] = transcript\n",
    "        \n",
    "    return output\n",
    "\n",
    "def create_indels2(matched_transcriptome, indels_per_read):\n",
    "    \n",
    "    output = np.empty(len(matched_transcriptome), dtype=\"object\")\n",
    "    \n",
    "    for i in range(len(matched_transcriptome)):\n",
    "        \n",
    "        new_read = matched_transcriptome[i]\n",
    "            \n",
    "            \n",
    "        change_indices = np.random.choice(range(len(new_read)),indels_per_read, replace=False)\n",
    "        \n",
    "        for change_index in change_indices:\n",
    "            if new_read[change_index]==\"1\":\n",
    "                new_read = new_read[:change_index]+\"4\"+new_read[change_index+1:]\n",
    "            \n",
    "            else:\n",
    "                insert=str(int(new_read[change_index])-1)\n",
    "                new_read = new_read[:change_index]+insert+new_read[change_index+1:]\n",
    "        \n",
    "        output[i]=new_read\n",
    "        \n",
    "    return output\n",
    "\n",
    "        \n",
    "class read_object2():\n",
    "    \"\"\"\n",
    "    create object to store the read we generate and its type...may expand class in the future to incorporate all\n",
    "    the read types we need as well as method to iterate over them.\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, variety, exome, read_length, indels_per_read = 7):\n",
    "        \n",
    "        self.group = variety\n",
    "        self.indels_per_read = indels_per_read\n",
    "        self.read_length = read_length\n",
    "        \n",
    "        #create indel read\n",
    "        if self.group ==\"indel\":\n",
    "            self.reads = create_indels2(create_matched_transcriptome2(exome, read_length),self.indels_per_read)\n",
    "        \n",
    "        #create good transcriptome read   \n",
    "        elif self.group ==\"good\":\n",
    "            self.reads = create_matched_transcriptome2(exome, read_length)\n",
    "        \n",
    "        #create bad transcriptome read\n",
    "        elif self.group == \"bad\":\n",
    "            self.reads = create_genome_exons2(int(len(exome)*len(exome[0])), read_length)\n",
    "\n",
    "class filter_object2():\n",
    "    \"\"\"\n",
    "    object to house the filter type and its attributes\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__ (self, exome, group_size, indices, set_filter = False, bloom =False, bloombits=1, fpr=0.01):\n",
    "        self.exome = exome\n",
    "        self.group_size = group_size\n",
    "        self.indices = indices\n",
    "        self.set_filter = set_filter\n",
    "        self.bloom = bloom\n",
    "        self.bloombits = bloombits\n",
    "        self.fpr=fpr\n",
    "    \n",
    "    def generate_filter(self, verbose = False):\n",
    "        \"\"\"\n",
    "        inputs:\n",
    "        set_filter: implement filter using has set rather than array\n",
    "        indices: generate the filter with partial keys beginning at these indices\n",
    "        filter object: filter array to be populated with seen reads, can experient with augmenting existing filters later\n",
    "        could have if/then stacked filter where we just keep grabbing larger partial keys (might be more efficient way\n",
    "        to access reads and score them)\n",
    "\n",
    "        output: filter with corresponding cells set to 1 for reads seen\n",
    "        \"\"\"\n",
    "        \n",
    "        #check to see if there are indices passed to filter obj\n",
    "        #assumes that all exons are the same size\n",
    "        if self.indices==False:\n",
    "            self.indices = range(len(self.exome[0]))\n",
    "           \n",
    "            \n",
    "        #use the set filter rather than array\n",
    "        if self.set_filter:\n",
    "            filter_array = set()\n",
    "            \n",
    "        if self.bloom == True:\n",
    "            filter_array = bloom.BloomFilter(capacity = self.bloombits, error_rate = self.fpr)\n",
    "            \n",
    "        keys_added=0\n",
    "        for i in range(len(self.exome)):\n",
    "\n",
    "            exon = self.exome[i]\n",
    "                \n",
    "            #initialize variable to index array of start indices\n",
    "            for j in self.indices:\n",
    "\n",
    "                #verify that we don't go out of range\n",
    "                if (j+self.group_size)>=len(exon):\n",
    "                    break\n",
    "\n",
    "                #add key to filter array\n",
    "                filter_array.add(exon[j:j+self.group_size])\n",
    "                keys_added+=1\n",
    "        print(f\"keys added: {keys_added}\")\n",
    "        self.filter_ = filter_array\n",
    "        \n",
    "        \n",
    "    def check_read(self, read, check_indices, quick_check = False, quick_thresh = 1, verbose = False):\n",
    "        \"\"\"\n",
    "        function checks to see if a read is present in the filter, utilizing self.filter_, group size and fwd_move\n",
    "        \n",
    "        read: transcriptome read to be mapped\n",
    "        check_indices: array of start positions that we want to check\n",
    "        quick_check: implement fast check to read...see if it passes threshold\n",
    "        quick_thresh: threshold to be used for quick check\n",
    "        \n",
    "        outputs:number of size(group size) partial keys that appear in our filter\n",
    "        \"\"\"\n",
    "        \n",
    "        #test all combos if check indices are not stipulated\n",
    "        if check_indices == False:\n",
    "            check_indices = range(len(read)-self.group_size+1)\n",
    "\n",
    "        #int to store number of partial reads that map\n",
    "        hits = 0\n",
    "            \n",
    "        #set threshold if we are using quick check\n",
    "        if quick_check:\n",
    "            thresh = len(check_indices)*quick_thresh\n",
    "                \n",
    "            #grab partial keys for all given start keys\n",
    "            for start_key in check_indices:\n",
    "\n",
    "                if verbose:\n",
    "                    print(f'key: {read[start_key:start_key+self.group_size]}')\n",
    "\n",
    "                #check set filter for key\n",
    "                if read[start_key:start_key+self.group_size] in self.filter_:\n",
    "                    hits+=1\n",
    "                    \n",
    "                #break if we have reached necessary threshold\n",
    "                if hits >thresh:\n",
    "                     return len(check_indices)\n",
    "        else:\n",
    "            #grab partial keys for all given start keys\n",
    "            for start_key in check_indices:\n",
    "\n",
    "                if verbose:\n",
    "                    print(f'key: {read[start_key:start_key+self.group_size]}')\n",
    "\n",
    "                #check set filter for key\n",
    "                if read[start_key:start_key+self.group_size] in self.filter_:\n",
    "                    hits+=1\n",
    "            \n",
    "        return hits\n",
    "    \n",
    "    def one_hit_check (self, read, check_indices):\n",
    "        \"\"\"\n",
    "        optimized check, breaks after one partial key hit that is in the filter, should be fastest\n",
    "        way to check\n",
    "        \n",
    "        returns: boolean true if partial appears in filter, false if not\n",
    "        \"\"\"\n",
    "        if check_indices == False:\n",
    "            check_indices = range(len(read)-self.group_size+1)\n",
    "            \n",
    "        for start_key in check_indices:\n",
    "\n",
    "            #check set filter for key\n",
    "            if read[start_key:start_key+self.group_size] in self.filter_:\n",
    "                        return True\n",
    "        return False\n",
    "\n",
    "       \n",
    "def test_filters2(read_groups, exome, group_size_array, read_indices_array, genome_indices_array, thresh_array, quick_check = False, set_filter = False, bloom=True, fpr=0.01):\n",
    "    \"\"\"\n",
    "    function to test filter params, going to begin by varying key size and forward moves\n",
    "    \n",
    "    inputs:\n",
    "    reads: array of references to arrays holding reads\n",
    "    exome: exome to create filter with \n",
    "    group_size_array: array holding group sizes to try\n",
    "    genome_indices_array: array of genome indices from which to build partial keys to populate filter\n",
    "    perc_thresh: percent threshold above which we want to count reads\n",
    "    read_indices_array: read_indices to try...need to figure out how to pass rules thru array to subsample\n",
    "    set_filter: whether to use hash set for filter vs array\n",
    "    spacing: set to 0 for now\n",
    "    quick_check: use the quick check method for check read\n",
    "    \n",
    "    output: dataframe with summary statistics for accuracy and time\n",
    "    \"\"\"\n",
    "    \n",
    "    #initialize output array variables\n",
    "    gs = []\n",
    "    gi = []\n",
    "    ri = []\n",
    "    good_percentage = []\n",
    "    good_variance = []\n",
    "    bad_percentage = []\n",
    "    bad_variance = []\n",
    "    good_thresh = []\n",
    "    bad_thresh = []\n",
    "    indel_thresh = []\n",
    "    indel_percentage = []\n",
    "    indel_variance = []\n",
    "    generate_time_array=[]\n",
    "    check_time_array=[]\n",
    "    exome_shape =[]\n",
    "    filter_full = []\n",
    "    read_lengths = []\n",
    "    num_reads=[]\n",
    "    thresh_used = []\n",
    "    \n",
    "\n",
    "    #loop over grid\n",
    "    for group_size in group_size_array:\n",
    "        for genome_indices in genome_indices_array:\n",
    "            for read_indices in read_indices_array:\n",
    "                for perc_thresh in thresh_array:\n",
    "                    \n",
    "                    print(\"a\")\n",
    "                    gs.append(group_size)\n",
    "                    gi.append(genome_indices)\n",
    "                    ri.append(read_indices)\n",
    "                    thresh_used.append(perc_thresh)\n",
    "                    \n",
    "                    #create filter\n",
    "                    if genome_indices == False:\n",
    "                        genome_read_size = len(exome[0])-group_size\n",
    "                        \n",
    "                    else:\n",
    "                        genome_read_size = len(genome_indices)\n",
    "                        \n",
    "                    bloombits=group_size*genome_read_size*len(exome)/10\n",
    "                    print(bloombits)\n",
    "                    filter_obj = filter_object2(exome, group_size, genome_indices, set_filter = set_filter, bloom = bloom, bloombits=bloombits, fpr=fpr)\n",
    "                \n",
    "                    #time how long it takes to create \n",
    "                    generate_beg = time.time()\n",
    "                    filter_obj.generate_filter()\n",
    "                    generate_end = time.time()\n",
    "                    print(\"generated filter\")\n",
    "\n",
    "                    #see how full the filter is\n",
    "#                     if set_filter:\n",
    "#                         filter_full_perc = len(filter_obj.filter_)/4**group_size\n",
    "                    \n",
    "#                     else:\n",
    "#                         filter_full_perc= sum(filter_obj.filter_)/4**group_size\n",
    "                  \n",
    "\n",
    "\n",
    "                    #append create time to output array, grabbing percentage of array that is filled\n",
    "                    generate_time_array.append(generate_end-generate_beg)\n",
    "\n",
    "                    #TODO generate read indices according to protocol\n",
    "\n",
    "                    #test all the transcriptome read objects and time them\n",
    "                    check_beg = time.time()\n",
    "                    for read_group in read_groups:\n",
    "\n",
    "                        #creat placeholder variable to store read by read results\n",
    "                        out_array = []\n",
    "                        for row in range(read_group.reads.shape[0]):\n",
    "\n",
    "                            #grab individual read and check it in the filter\n",
    "                            read = read_group.reads[row]\n",
    "                            out_array.append(filter_obj.check_read(read, read_indices,quick_check=quick_check,quick_thresh=perc_thresh))\n",
    "\n",
    "\n",
    "                        #establish number of partial keys we tried\n",
    "                        if read_indices == False:\n",
    "                            pos = len(read_group.reads[0])-group_size+1\n",
    "\n",
    "                        else:\n",
    "                            pos = 0\n",
    "                            for i in read_indices:\n",
    "                                if i+group_size<len(read_group.reads[0]):\n",
    "                                    pos+=1\n",
    "\n",
    "                        #calculate ave percentage of hits\n",
    "                        percentage_array = [i/pos for i in out_array]\n",
    "                        prec = sum(percentage_array)/len(percentage_array)\n",
    "\n",
    "                        #calcualte percentage of reads above thresh\n",
    "                        thresh=0 \n",
    "                        for i in percentage_array:\n",
    "                            if i >=perc_thresh:\n",
    "                                thresh+=1\n",
    "\n",
    "                        thresh = thresh /len(percentage_array)\n",
    "\n",
    "\n",
    "                        #calculate variance of num hits\n",
    "                        var = np.var(out_array)\n",
    "\n",
    "\n",
    "                        if read_group.group == \"indel\":\n",
    "                            indel_percentage.append(prec)\n",
    "                            indel_variance.append(var)\n",
    "                            indel_thresh.append(thresh)\n",
    "\n",
    "                        elif read_group.group == \"good\":\n",
    "                            good_percentage.append(prec)\n",
    "                            good_variance.append(var)\n",
    "                            good_thresh.append(thresh)\n",
    "\n",
    "                        else:\n",
    "                            bad_percentage.append(prec)\n",
    "                            bad_variance.append(var)\n",
    "                            bad_thresh.append(thresh)\n",
    "\n",
    "                    check_end = time.time()\n",
    "                    \n",
    "                    check_time_array.append(check_end-check_beg)\n",
    "                    exome_shape.append(exome.shape[0])\n",
    "                    filter_full.append(0)\n",
    "                    read_lengths.append(len(read_groups[0].reads[0]))\n",
    "                    num_reads.append(read_groups[0].reads.shape[0])\n",
    "\n",
    "\n",
    "        out_df = pd.DataFrame({\"Group Size\":gs,\"Genome Indices\":gi,\"Read Indices\":ri, \"Good Percentage\":good_percentage,\n",
    "                              \"Bad Percentage\":bad_percentage, \"Indel Percentage\":indel_percentage, \"Good Variance\":good_variance,\n",
    "                              \"Bad Variance\":bad_variance, \"Indel Variance\":indel_variance, \"Good Thresh\":good_thresh,\n",
    "                              \"Bad Thresh\":bad_thresh, \"Indel Thresh\": indel_thresh, \"Create Time\": generate_time_array,\n",
    "                              \"Check Time\": check_time_array,\"Exome Size\":exome_shape, \"Read Lengths\": read_lengths, \"Num Reads\":num_reads, \n",
    "                              \"Filter Full Percentage\":filter_full, \"Threshold Used\":thresh_used})\n",
    "        \n",
    "    return out_df, filter_obj  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-02T04:44:25.164414Z",
     "start_time": "2022-08-02T04:44:25.144413Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1333222114', '1131143441', '3122423314', '4214114242'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a=create_genome_exons2(40,10)\n",
    "b=create_matched_transcriptome2(a,10)\n",
    "c=create_indels2(b,4)\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-02T04:44:26.003437Z",
     "start_time": "2022-08-02T04:44:25.994437Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1333222114', '3332221141', '3322211411', '3222114113',\n",
       "       '2221141131', '2211411311', '2114113114', '1141131143',\n",
       "       '1411311434', '4113114344', '1131143441', '1311434413',\n",
       "       '3114344131', '1143441312', '1434413122', '4344131224',\n",
       "       '3441312242', '4413122423', '4131224233', '1312242331',\n",
       "       '3122423314', '1224233144', '2242331442', '2423314421',\n",
       "       '4233144214', '2331442141', '3314421411', '3144214114',\n",
       "       '1442141142', '4421411424', '4214114242'], dtype=object)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-02T04:44:26.861711Z",
     "start_time": "2022-08-02T04:44:26.843712Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1333111414', '2322211131', '2312214414', '2121414113',\n",
       "       '2114141134', '1211341211', '1114143444', '1144134132',\n",
       "       '4441341334', '4112443344', '4131433444', '1211424343',\n",
       "       '3113333134', '1442444311', '4333413422', '4243431223',\n",
       "       '3434341242', '3412422422', '4124224222', '1311141321',\n",
       "       '3111423343', '1123232444', '2142224442', '2412244421',\n",
       "       '4223143144', '1321442431', '2314314411', '3134243414',\n",
       "       '4442444442', '4421314323', '3114114141'], dtype=object)"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-02T04:47:57.526163Z",
     "start_time": "2022-08-02T04:46:20.455302Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "bad2 (30000, 100)\n",
      "good2 (2999901, 100)\n",
      "indel2 (2999901, 100)\n",
      "total 6029802\n",
      "exome length 100\n",
      "exon length 30000\n",
      "time to generate: 97.0588526725769\n"
     ]
    }
   ],
   "source": [
    "a= time.time()\n",
    "exome2 = create_genome_exons2 (int(3e6),int(3e4))\n",
    "#bad_transcriptome = create_transcriptome(80,20)\n",
    "#good_transcriptome = create_matched_transcriptome(exome, 20)\n",
    "#indels = create_indels(good_transcriptome,1)\n",
    "print(1)\n",
    "bad2 = read_object2(\"bad\", exome2, 100)\n",
    "print(2)\n",
    "good2 = read_object2(\"good\",exome2, 100)\n",
    "print(3)\n",
    "indel2 = read_object2(\"indel\",exome2, 100)\n",
    "b=time.time()\n",
    "\n",
    "print(f\"bad2 {len(bad2.reads), len(bad2.reads[0])}\")\n",
    "print(f\"good2 {len(good2.reads),len(good2.reads[0])}\")\n",
    "print(f\"indel2 {len(indel2.reads),len(indel2.reads[0])}\")\n",
    "print(f\"total {len(bad2.reads)+len(good2.reads)+len(indel2.reads)}\")\n",
    "print(f\"exome length {len(exome2)}\")\n",
    "print(f\"exon length {len(exome2[0])}\")\n",
    "print(f\"time to generate: {b-a}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-02T05:39:28.474220Z",
     "start_time": "2022-08-02T05:38:37.510853Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "a\n",
      "5996000.0\n",
      "keys added: 2998000\n",
      "generated filter\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Group Size</th>\n",
       "      <th>Genome Indices</th>\n",
       "      <th>Read Indices</th>\n",
       "      <th>Good Percentage</th>\n",
       "      <th>Bad Percentage</th>\n",
       "      <th>Indel Percentage</th>\n",
       "      <th>Good Variance</th>\n",
       "      <th>Bad Variance</th>\n",
       "      <th>Indel Variance</th>\n",
       "      <th>Good Thresh</th>\n",
       "      <th>Bad Thresh</th>\n",
       "      <th>Indel Thresh</th>\n",
       "      <th>Create Time</th>\n",
       "      <th>Check Time</th>\n",
       "      <th>Exome Size</th>\n",
       "      <th>Read Lengths</th>\n",
       "      <th>Num Reads</th>\n",
       "      <th>Filter Full Percentage</th>\n",
       "      <th>Threshold Used</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>20</td>\n",
       "      <td>False</td>\n",
       "      <td>[0, 20, 40, 60, 80]</td>\n",
       "      <td>1.25</td>\n",
       "      <td>0.001792</td>\n",
       "      <td>1.249844</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.035782</td>\n",
       "      <td>0.003125</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.001433</td>\n",
       "      <td>0.999875</td>\n",
       "      <td>13.386922</td>\n",
       "      <td>37.105834</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>30000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Group Size  Genome Indices         Read Indices  Good Percentage  \\\n",
       "0          20           False  [0, 20, 40, 60, 80]             1.25   \n",
       "\n",
       "   Bad Percentage  Indel Percentage  Good Variance  Bad Variance  \\\n",
       "0        0.001792          1.249844            0.0      0.035782   \n",
       "\n",
       "   Indel Variance  Good Thresh  Bad Thresh  Indel Thresh  Create Time  \\\n",
       "0        0.003125          1.0    0.001433      0.999875    13.386922   \n",
       "\n",
       "   Check Time  Exome Size  Read Lengths  Num Reads  Filter Full Percentage  \\\n",
       "0   37.105834         100           100      30000                       0   \n",
       "\n",
       "   Threshold Used  \n",
       "0            0.01  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "threes=list()\n",
    "twos=list()\n",
    "fours = list()\n",
    "fives = list()\n",
    "sixes = list()\n",
    "sevens = list()\n",
    "twos2=list()\n",
    "twosfin=list()\n",
    "\n",
    "\n",
    "i=0\n",
    "for i in range(int(3e5)):\n",
    "    if i%2==0:\n",
    "        twos.append(i)\n",
    "        twos2.append(i+1)\n",
    "        \n",
    "    if i%3==0:\n",
    "        threes.append(i)\n",
    "        \n",
    "    if i%4==0:\n",
    "        fours.append(i)\n",
    "        \n",
    "    if i%5==0:\n",
    "        fives.append(i)\n",
    "        \n",
    "    if i%6==0:\n",
    "        sixes.append(i)\n",
    "    if i%7==0:\n",
    "        sevens.append(i)\n",
    "\n",
    "for i in range(len(twos2)):\n",
    "    if i%2==0:\n",
    "        twosfin.append(twos2[i])\n",
    "    else:\n",
    "        twosfin.append(twos[i])\n",
    "        \n",
    "genome_indices = [False]\n",
    "read_indices = [[0,20,40,60,80]]\n",
    "partial_key_size = [20]\n",
    "thresholds=[0.01]\n",
    "\n",
    "a, b=test_filters2([bad2, good2, indel2],exome2, partial_key_size, read_indices ,genome_indices, thresholds,quick_check=True, set_filter = False, bloom=True, fpr=0.01)\n",
    "#a.to_csv(\"genome_index_300mm_5.30.csv\")\n",
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-02T05:57:14.280869Z",
     "start_time": "2022-08-02T05:57:14.264871Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5153244.0"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samps = ((429454)-17)*12\n",
    "expected_unique=(1-(1-1/samps)**(4**50))*samps\n",
    "expected_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "a, b=test_filters2([bad2, good2, indel2],exome2, partial_key_size, read_indices ,genome_indices, thresholds,quick_check=True, set_filter = False, bloom=True, fpr=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-17T01:03:23.304998Z",
     "start_time": "2022-06-17T01:03:23.191574Z"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD4CAYAAAD8Zh1EAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAlw0lEQVR4nO3deXyV5Z338c8vKwQCYUlYEnZZZAcDarUqaisWKqC1BWfa6TbWTu3mTLfptDN2nk6fttOZbnR8nNq9BQXZtCjW1t0qJISwiIHIkuQkZAESQkLWcz1/nNCmFCXLdXJyzvm+Xy9ecs65uc4vt8n33Lnu+/rd5pxDRESiX0KkCxARET8U6CIiMUKBLiISIxToIiIxQoEuIhIjkiL1xiNHjnQTJ06M1NuLiESl/Pz8Gudc5sVei1igT5w4kby8vEi9vYhIVDKz42/2mqZcRERihAJdRCRGKNBFRGKEAl1EJEYo0EVEYsQlA93MfmJmVWa2/01eNzP7vpkVm9leM1vov0wREbmUrhyh/wxY+hav3wpM7fhzN/A/vS9LRES665LXoTvnnjeziW+xyQrgFy7Uh/cVM8swszHOuQpfRXZWdKKe3+4tD8fQ3TYlazAr5mf36N+eamjh+UPVZKQlc8P0rB6N0R50PH2wkhGDUsidOLxHYzjnePpgFWMzBjBr7NAejSEi/YOPhUXZQGmnx2Udz/1VoJvZ3YSO4hk/fnyP3qy46iw/eKa4R//WJ+cgOdG6HOjBoGNfoI5ni6p5pqiKwrJanIOxQwfw8pdu6tZ7N7a0sTG/jB+/cJSSU43MyxnK1nuv7fbXsD9Qx79tO0De8dPcNCOLhz64qNtjiEj/4SPQ7SLPXfSuGc65B4EHAXJzc3t0Z41lc8ewbO6ynvxTr77zVBFrL/HBcrqhhecPV/NcUTXPHarmZEMLZjAvJ4PP3DSNvOOnOFRZ3+X3rDnbzC9ePsYvXjlObWMrC8ZnkJRotAW7tytPN7Twn08VsW5nCRlpKQxLS+72GJ0FO35TmD46nQkjBvV4HBHpHR+BXgaM6/Q4B+gfcyJ9LBh0HCg/wzNFVTxbVMWe0lqCDoalJXP9tExumJ7FddMyGT4oBYAvPrq3S4F+pPos//vCUR7dXUZre5CbLx/Fx66bTO7E4Xz057uoqGvqUn3tQcdvdpbwnaeKqG9q4wNXT+SzN0/jAz/d2eOveV9ZHV/Zup89pbWsWTyOb9w+t8djiUjv+Aj0bcC9ZrYeuBKoC9f8eX9U19jK84erebaomucOVVFztgWAeTlDuffGqSyZnsncnAwSEy72i8xbyzt2igefP8LvDlaSnJjAHQtz+OjbJzElc3C3x9p17BT/uvUAr1Wc4cpJw7l/xSxmjB7S7XHOO93Qwrc7jvJHDEplYHIire26naFIJF0y0M1sHXADMNLMyoB/BZIBnHMPANuBdwHFQCPwoXAV298EHSz496cIOhg6MJnrpmWyZHom103LZOTg1B6N2R50/O61Sh58/g12l9SSkZbMvUsu4wNXTyQzvftjVp5p4hvbD7JlTzljhg7gB2sWsHzuGMy6/wEDod9CHs4r5VtPvs6ZpjY++LaJfPYd07j1uy/0aDwR8acrV7msucTrDviEt4qixKKJw1k8aThXThrODdOzmD+uZ0fh5zW1trMxv4yHXjzK0ZoGxg0fyP23zeLO3BzSUrr/i1RLW5CfvHSUH/z+MK3tjk8smcInllzWo7HOKyyt5atb91NYVsfiiaGj/MvH9PwoX0T8ilj73Gh33bTQkXhvtbQF+d7Th/nFH49xsqGFeTlDWXvXQpbOHt3jD4hni6r42mOvcaSmgZsvz+Iry2f26mTl6YYWvrWjiPW7Shg5OJXvvm8+K+aP7fFRvoiEhwI9wk43tvLfTx/iphlZ/P11k7ly0vAeB2XJyUa+9vhrPH2wkkkjB/HTDy1iSQ+vcYfQ9M/6XSV8e0foJOpHrpnEp2+eSvqA5B6PKSLho0CPoBXzs0lLSWLN4nFMHZXe43Ga24J856ki/t/zR0hKML6wdAYfvnYiqUmJPR6zoOQ0X916gH2BOq6aPJyvrZjNtF7UKCLhp0CPoKunjODqKSN6PU5x1Vl+8IdiVswfy5duvZzRQwf0eKyTZ5v51pNFPJxXyqghqXx/zQLe3YuTqCLSdxToUW7+uAxON7byhaUzWDypZ8v/AYLO8cs/HuM/nzpEQ3Mbd183mU/dNJXBqfoWEYkW+mmNcvfeOJV7b5za63FeOFzDC4dreNuUEdx/26xeTQGJSGQo0IXRQ1KpHDKAf1l+OcvmaHpFJFop0IW1dy3EzHp1Hb0PgdpzjBiUwoDknp/MFYlnumORkJSYEPEwr21s4Z3/9RwPvXg0onWIRDMFuvQLmwsCNLS0U9/UFulSRKKWAl0izjnHup0lkS5DJOop0CXidpfUcqjybKTLEIl6CnSJuHU7SxiUkhjxeXyRaKdAl4iqO9fK43vLuW1+tpdAP9vcRqgBqEj8UaBLRG3bE6CpNchdi3t2j9nOquqbuOo/fs+2wri8YZaIAl0ixznHr18tYXb2EObkDO31eI8VVnC2uY3q+mYP1YlEHwW6RExhWR2vn6hn9aLeH50DbCkIeBlHJFop0CVi1u8sYWByIivmj+31WMVV9ewL1HmoSiR6KdAlIuqbWtlWWM5t88Z6uWHGZh2diyjQJTK2FZbT2NLO6sXjej1WMOjYUlDOwvEZvS9MJIop0CUi1u8sZcbodOaPy+j1WLuOnSJQe45VC7J7X5hIFFOgS5/bV1bHvkAdd1053kur3s0FAdJSEnnnrNEeqhOJXgp06XPrdpUwIDmBFfN7f0Td1NrOb/dVsHTWaAamqO2uxDcFuvSphuY2tu0pZ9mcsQwd2PuToX94vYr6pjZWarpFRIEufevxveWcbW5jjYeToRCabslMT+Way0Z6GU8kminQpU/9ZmcpU7MGc8WEYb0e63RDC88WVbFi3lg19hJBgS596LXyMxSW1rJmsZ+ToY/vq6C13bFqob/plpqzzaxY+xIHK854G1OkryjQpc+s31VCSlICt3sK4C0FAaaNGszMMUO8jAfwaH4ZhaW1HKqs9zamSF9RoEufONfSzuaCAO+aPZqMtJRej3f8ZAP5x0+zakGOl6N9CDUL25hf5mUskUhQoEuf+O2+Cuqb2ljjoU0uwJaCcszw0gfmvMKyOg5X6c5JEr0U6NIn1u0sYXLmIBZPGt7rsZxzbC4o46pJIxibMdBDdSEb8kq9jSUSCQp0CbtDlfXkHz/NmkV+TobuKa3l2MlGr0v9m1rb2VZYzjwPfdlFIqVLgW5mS82syMyKzeyLF3l9qJk9ZmaFZnbAzD7kv1SJVut2lpCcaN5Ohm4uCJCalMDSOf6W+j/1WiX1TW2854ocb2OK9LVLBrqZJQJrgVuBmcAaM5t5wWafAF5zzs0DbgC+Y2a9P/MlUa+ptZ1NuwPcMms0Iwan9nq81vYgjxWWc/PMUQzx0Hb3vA15pWRnDOTqKVqgJNGrK0foi4Fi59wR51wLsB5YccE2Dki30O/Tg4FTQJvXSiUqPbn/BHXnWr3cMxTg+UPVnG5sZZWHPjDnVdSd48XiGu5YmI3WJ0k060qgZwOdzxaVdTzX2Q+By4FyYB/waedc8MKBzOxuM8szs7zq6uoelizR5Dc7S5gwIo2rJo/wMt6mggDD0pK5fnqml/EANu0O4BzcoekWiXJdCfSLHbO4Cx7fAuwBxgLzgR+a2V+t9nDOPeicy3XO5WZm+vuBlP6puOosO4+eYvWi8SR4OPQ909TK069V8u55Y0lO9HM+3znHhrxSFk8azoQRg7yMKRIpXfmpKAM6d1LKIXQk3tmHgE0upBg4CszwU6JEq4d3lZCUYN5OND657wTNbUGvV7fkHz/NsZON3Kmjc4kBXQn0XcBUM5vUcaJzNbDtgm1KgJsAzGwUMB044rNQiS7Nbe1szC/jHTNHkZne+5OhELq6ZdLIQV7ucnTehrwy0lISedecMd7GrGts5d+2HeBss04jSd+6ZKA759qAe4EdwEHgEefcATO7x8zu6djs34G3mdk+4PfAF5xzNeEqWvq/HQcqOd3Y6m1laHntOV45epKV87O9LfVvbGnj8b3lvGvOGAalJnkZE+DXO4/zs5ePsbe01tuYIl3Rpe9i59x2YPsFzz3Q6e/lwDv9libRbP3OEnKGDeRaT33Kt+4pxzlYucDfUv8n95+goaXd63RLMOhYv1MrTiUytFJUvDta08DLb5xk9aJxXk6Gnl/qv3B8htcTlxvyyhg/PM1LO4Lz/njkJCWnGr2NJ9IdCnTxbv2uEhITjDtz/dyV6LWKMxyqPMuqhf6OpEtPNfLHIyd5zxX+ujVCaFWsSKQo0MWrlrYgj+aXceOMLEYNGeBlzC0FAZITjeUeT1w+ursMM7/Xnp9qaOGpA5XMGuuvP7tIdyjQxaunD1ZSc7bF28rQ9qBj655ybpiexbBBfrpJBIOhvufXTBlJtsdujZt2l9HSHmS1h6+97lwr//v8EYLBC5d8iLw5Bbp4tW5nCWOHDuC6aX4Wjr38Rg1V9c1erz1/5ehJyk6f89qIyznHb3aWsHB8BtOyBvd6vJ+8eJSvbz/IkRr1Z5euU6CLN6WnGnnhcA3vXTTO202bN+8OkD4giRtnZHkZD2BjfhnpqUncMstft8Zdx05zpLrBy9G5c44tewIA6ABdukOBLt68evQUCQbv9XQytLGljScPnGDZnDEMSE70MubZ5jae2HeC5fPGMjDFz5gQukwzPTWJ5XN7P89fUFrL8ZO6Uka6T4EuXi2ZnuXtLkJPHaiksaXd63TLb/eWc6613et0S11jK7/dV8GKBWNJS+n9AqXNuwMeqpJ4pEAXr3xMOZy3uSBAdsZAFk30d534xvwyJmcOYuH4DG9jbtkToLktyOpFvf/aW9qCPL63nEEef3uQ+KFAF29GDUlliae2tlX1TbxwuJoV88d6WZwEoQVPu46d5s4rxnm79tw5x7qdJczJHsrs7N7fvu65jn7vS2f7u0RT4ocCXbxYPm8Mn7tlBkme2to+VlhB0OF1umVjfikJhrdb4UHo/qavn6hn9WI/5w22FAQYPiiF66bpzknSff46Eklc+9Ktl3sdb0tBgNnZQ5g6Kt3LeO1Bx6bdAa6blultwRPA+p2lpKUkctu83veYOdPUyu8OVrJm0TiSEnSsJd2n7xrpd4qr6tkXqGPVAn8nLl8qrqGirok7r/BzJA2hK2Ye21vOu+eOJd3D/U2f2FdBS1vQa4sDiS8KdOl3NhcESDB49zx/88gb8ssYOjCZm2f6u559255yGlvavU23nO/3Pi+n93PxEp8U6NKvOOfYUlDO26dmkpXuZ2qk7lwrOw6cYMX8saQmebz2fFcJM0ane7nhRqD2HK8cOeW137vEHwW69Cu7jp0iUHvO68nQxwrLaWkLep1u2R+oY29ZHasX+bliZmvHylCfX3dLW5B9ZXXexpP+T4Eu/cruklrSUhJ556xR3sbcmF/GjNHpzM721wVx/a4SUpMSvMzzO+fYvDvAFROGMX5EmofqQtY+U8zKH71EfVOrtzGlf1OgS7+zdNZoLysuIXSCdU9prde+540tbWwtKGfZnDEMTev9ydDXKs5wuOosKz0enZ/vKNkedLS2qyFMvFCgS7/jM9g25JWRlGBex/zt3grqm9u8rYrdvNt/v/c/HjlJoPact/EkOijQpV/JTE/lGk/3IW1rD7KpIMCSGVmMHJzqZUyA9btKmZI5iEUTh/V6rPagY2uh337vABvydF/TeKRAl35jxKAU3pfrr/Xu84erqa5v9tqI61BlPfnHT7N60XgvUzgvv1FDdX0zt3v8DeJMUytP7D/BgGT9eMcbrRSVfuN3913PQE9tciE03TJiUIrXXurrdpaQnGje2gec7/e+xGONv91bQXNbkNsXZLOpQJ0b44k+wqXfGJya5O3o/HRDC08frGTlgmySPfWXaWptZ3NBgFtmjWaEhymccPR7h9B0y9SswczVAqW4o0CXmLR1T4DWdud1umXHgRPUNrayxtPJ0HD0e3+j+iy7S/xe1SPRQ4EuMWlDfhmzs4dw+Rh/156v21nC+OFpXD15hJfxwtXvPTHBvH5ISPRQoEvMOVhRz4HyM15Xhh6pPssrR07xvkXjvPRnr65v9t7vPdRRsozrp2WS5bGjpEQPBbrEnCf2V5CSmOClpe15D+8qJTHBuNPTFM62wnLv/d5fOFxN5ZlmbzVK9FGgS8xpbGnn5pn+rutuaQuyMb+Mm2ZkeTvy9d3vHULTTBlpydx4ub8rZiS6KNAlJvmcbnn6YCUnG1pYc6Wfk6Hn+72vnO/v6LyusZXfHahk5fxsrx0lJboo0CXmZKWn8vap/m7htm5nCdkZA7luqp/7pZ7v937bfH9TQtsKA7S0B71e1SPRR4EuMSMpwUhMMG5fmOPt3qalpxp5sbiGO3NzvFwjHwyG+r1f67HfO/y5o+Sssf6u6pHoo0CXmJGWksQjH7uKz9w81duYj+SVYsB7c/1M4eQdP02g9pzXpf6HKuspLKvjzlw/vdklenUp0M1sqZkVmVmxmX3xTba5wcz2mNkBM3vOb5kiXXPFhOHeVl22tQd5JK+U66dlMjZjoJcxNxeUhaXfe1KCsdLjFI5Ep0sGupklAmuBW4GZwBozm3nBNhnAj4DbnHOzgDv9lyrSt54pCl0G6GtlaFNrO4/vreAWj/3eW9uDbNod4MYZWV7aEUh068oR+mKg2Dl3xDnXAqwHVlywzV3AJudcCYBzrspvmSJ9b/3OErLSU70193rm9Srqm9q8Xnv+XFE1NWf9dpSU6NWVQM8GOjdXLut4rrNpwDAze9bM8s3sAxcbyMzuNrM8M8urrq7uWcUifaCirolniqq4M9ffCdbNBQEy01N52xQ/rQMgNN0ycnCK126NEr268p16sbMsF97TKgm4AlgG3AJ8xcym/dU/cu5B51yucy43M9PPJWAi4bAhr5Sgg/fl+pluqW1s4ZmiKm6bN9bbB8SphhZ+/3ro2nNfHSUlunVlIq8M6HyKPwcov8g2Nc65BqDBzJ4H5gGHvFQp0sfeqG7g2stGertp8+N7K2htd16nW/7UUTJX0y0S0pWP9V3AVDObZGYpwGpg2wXbbAXebmZJZpYGXAkc9FuqSN9avdjfatMtBQGmZg32ep34hrwy5mQPZcZoXXsuIZcMdOdcG3AvsINQSD/inDtgZveY2T0d2xwEngT2AjuBHzvn9oevbJHwGjEohXfOHO1lrJKTjeQdP83KBdnerhM/UF7HaxVndDJU/kKXrp1yzm0Htl/w3AMXPP428G1/pYn0vfOBe8cVOaQk+ZmX3rIndBu4lR6nWzbml3nvKCnRT/cUFelkwvA0PnfLdFYv8jPd4pxjS0GAKycNJ9vT4qSWtiBb95R77SgpsUGnxkU6SUgwPrHkMm+LdPaW1XGkpsHrydA/vF7FqYYWrx0lJTYo0EXCaHNBgJSkBG6dM8bbmBvzS713lJTYoEAXCZPWoOOxwnJuvjyLoQOTvYxZXd/MM0XVrFqY7e16dokd+o4QCZMXDlVzsqGFVQv8XYmypSBAe9DpNnNyUQp0kTDZVlhORloy10/zsyraOcfG/DLmj8vgsix/t66T2KFAFwmTqvpmls8d4+3yx32BOooq67lTK0PlTSjQRcLI53TLhrwyUpMSWD5X157LxSnQRcJk/PA0Fo7P8DJWU2s72wrLuWXWaG8nWCX2KNBFwsTnUv+nD1ZSd65VS/3lLSnQRTybPjqdd84cxV2e7nQEoaX+Y4YO4JrLdO25vDkt/RfxLCMthQc/kOttvBN1TTx/qJqP3zCFxATdBFrenI7QRfq5zQUBgg7eo6X+cgkKdJF+zDnYkF/KoonDmDRyUKTLkX5OgS7Sj+0pPc2R6gadDJUuUaCL9GMb8soYmJzIMl17Ll2gQBfpx/KOn+bW2aMZnKrrF+TSFOgi/ZxuAi1dpUAX6cdyhg3kqkkjIl2GRAkFukg/dsfCHBJ07bl0kQJdpB+aMCKNccMH8l5P9zaV+KAzLSL90Ozsobzw+RsjXYZEGR2hi4jECAW6iEiMUKCLiMQIBbqISIxQoIuIxAgFuohIjFCgi4jECAW6iEiMUKCLxKhBHR0adx8/HeFKpK90KdDNbKmZFZlZsZl98S22W2Rm7Wb2Hn8likhPLJ87humj0vnS5n2camiJdDnSBy4Z6GaWCKwFbgVmAmvMbOabbPdNYIfvIkWk+wYkJ/Lf75tPbWML/7JlH865SJckYdaVI/TFQLFz7ohzrgVYD6y4yHafBB4FqjzWJyK9MHPsED77jmls33eCLXsCkS5HwqwrgZ4NlHZ6XNbx3J+YWTawCnjAX2ki4sPHrptC7oRhfHXrAcprz0W6HAmjrgT6xZoxX/i723eBLzjn2t9yILO7zSzPzPKqq6u7WKKI9EZigvGd986jPej4pw2FBIOaeolVXQn0MqBzU+YcoPyCbXKB9WZ2DHgP8CMzW3nhQM65B51zuc653MzMzJ5VLCLdNmHEIL6yfCYvv3GSn718LNLlSJh0JdB3AVPNbJKZpQCrgW2dN3DOTXLOTXTOTQQ2Av/gnNviu1gR6bnVi8Zx04wsvvnk6xRX1Ue6HAmDSwa6c64NuJfQ1SsHgUeccwfM7B4zuyfcBYqIH2bGN+6YQ1pKIp99uJDW9mCkSxLPLFKXMuXm5rq8vLyIvLdIPHtyfwX3/Go3n7rxMu575/RIlyPdZGb5zrnci72mlaIicWbp7DHcvjCbtc++QUGJVpHGEgW6SBz6t9tmMXrIAO57pJDGlrZIlyOeKNBF4tCQAcl8+865HK1p4BvbX490OeKJAl0kTr1tykg+cu0kfvnKcZ47pHUhsUCBLhLHPnfLdKZmDeZzGwqpbVQDr2inQBeJY+cbeJ1qaOFftuyPdDnSSwp0kTg3O3son7l5Ko/vrWCrGnhFNQW6iHDP9VNYMD6Dr2zZT0WdGnhFKwW6iJCUmMB/v3c+re2Oz2/cqwZeUUqBLiIATBw5iC8vu5wXDtfwy1eOR7oc6QEFuoj8yd9cOZ4bpmfyjScO8kb12UiXI92kQBeRPzEzvnXHXAYkJ3Lfw3vUwCvKKNBF5C9kDRnA11fOobCsjrXPFEe6HOkGBbqI/JVlc8ewcv5YfvCHYgpLayNdjnSRAl1ELur+FbPJSk/l8xv3RroU6SIFuohc1NCBydy+MJvDurtR1FCgi8ibSrCL3SNe+isFuohIjFCgi4jECAW6iEiMUKCLSFhtLijju08finQZcUGBLiJhs62wnPseKWTdzpJIlxIXFOgiEhbPFlVx38N7cGrc2GcU6CLiXf7x03z8V7uZNiqdZXPGRLqcuKFAFxGvik7U8+Gf7WLUkFR+/uHFpA9IinRJcUOBLiLelJ5q5P0PvcqA5AR++ZEryUxPjXRJcUUfnSLiRXV9M+9/6FWa24I88rGrGTc8LdIlxR0doYtIr51pauXvfrKTyjPN/OSDi5g+Oj3SJcUlBbqI9EpTazsf/Xkeh6vqeeD9V3DFhGGRLiluacpFRHqsrT3Ivb/Zza5jp/je6gVcPy0z0iXFNR2hi0iPBIOOLzy6j6cPVvG1FbO5bd7YSJcU9xToItJtzjm+vv0gj+4u4753TOP9V02IdElCFwPdzJaaWZGZFZvZFy/y+t+Y2d6OPy+b2Tz/pYpIf/GjZ9/goReP8sG3TeSTN14W9vd7qbiGFw5Xh/19ot0l59DNLBFYC7wDKAN2mdk259xrnTY7ClzvnDttZrcCDwJXhqNgEYms37xawrd3FLFy/li+unwmFsabYDS1tvN/n3idn718jDnZQ3n7VM3Rv5WunBRdDBQ7544AmNl6YAXwp0B3zr3caftXgByfRYpI/7B9XwVf3rKPG2dk8e0755GQEL4wP1xZzyfXFfD6iXoGpyYRVFOYS+rKlEs2UNrpcVnHc2/mI8ATF3vBzO42szwzy6uu1q9PItHkhcPVfHp9AVeMH8bauxaSnBieU3DOOX796nHe/cMXqa5v5qcfXMRVk4eH5b1iTVf+j1zsI/iiH5VmtoRQoH/hYq875x50zuU653IzM/Wrk0i02FNay8d+mc+UzME89MFFDExJDMv71Da28PFf7ebLm/ezaOJwnvjM21kyIyss7xWLujLlUgaM6/Q4Byi/cCMzmwv8GLjVOXfST3kiEmlBBx/86U5GDk7lFx9ezNCByWF5n1eOnOSzD++h5mwz//yuGXz02slhndKJRV0J9F3AVDObBASA1cBdnTcws/HAJuD9zjndmkQkxiQnJvCrj1xJ1pAB3sduaw/yvd8f5ofPFDNxxCA2ffwa5uQM9f4+8eCSge6cazOze4EdQCLwE+fcATO7p+P1B4CvAiOAH3Wc8W5zzuWGr2wR6QtZQwaQkZbMLz68mPEj/DfbKj3VyKfXF7C7pJb3XJHD/bfNYlCqFrD3VJf2nHNuO7D9guce6PT3jwIf9VuaiETa+6+awPtyx5GS5P8E6LbCcr68aR8A31+zQCtNPdBHoYi8Jd9h3tDcxr9uO8DG/DIWjM/g+6sXqNWuJwp0Eekz+wN1fHJdAcdONvDJGy/jUzdNDdvlj/FIgS4iYRd08L/PH+FbO15n5OBU1v39VVw1eUSky4o5CnQRCbvq+ma+vv0gt8waxTfvmEtGWkqf13Ck+iwZaSkMH9T3791XFOgiElaDU5NITUrgq++eyV2Lx4e198uFqs40sXVPOZsLArxWcYZlc8ew9q6Fffb+fU2BLiJh9U+3TOeeG6YwcnDf3DC6obmNHQdOsLkgwEvFNQQdzMsZysjBKZxtauuTGiJFgS4iYTUgOZEByeFpFXBeW3uQl944yebdZew4UMm51nZyhg3kE0suY+WCbKZkDmbF2pfCWkN/oEAXkajknONA+Rk2FwTYVlhOdX0zQwYksWphNqsWZJM7YVifTu/0Bwp0EYkqgdpzbN0TYPPuAIerzpKcaNw4I4tVC7JZMiOL1KTw/jbQnynQRSQqVJ5pYvWDf+SVI6cAyJ0wjK+vms2yOWP67KqZusZW0lIT++218wp0Een3BqYkUXO2hfQBzdz3jmmsnJ8dlt4yF3O2uY0d+ztOsr5Rw91vn8yX3nV5n7x3dynQRaTf++rymdxz/WRmjhnSJ/Pibe1BXiyuYXNBgKc6TrKOGz6Q1KQETja0hP39e0qBLiL9XmZ6Kpnp4b3s0TnH/sCfT7LWnG1m6MBkbu84yXrFhGFc+81nwlpDbynQRSSulZ1u/NPio+Kqs6QkJoROsi7M5obpmVF1klWBLiJxp+5cK0/sq2BzQYBXj4ZOsi6aOIz/WDWHZXPGMDQtPHdlCjcFuojEjbLTjfzDr/N5+mAVLW1BJo8cxD++YxorF2THRAtfBbqIxIWUROON6gZqG1u5a/F4Vi3IZm7O0JhafKRAF5G4cP9ts6mqb+Kay0b22+vIe0uBLiJxYebYIcxkSKTLCKvY/JgSEYlDCnQRkRihQBcRiRGaQxcR6SIz2FIQ4Mn9J3o1xv9ZOZsV87M9VhaiQBcR6aIv3Xo5u0tO92qMn750lIMV9ayY76emzhToIiJdtGzuGJbNHdOrMX75ynFP1fw1zaGLiMQIBbqISIxQoIuIxAgFuohIjFCgi4j0oaWzRjNjdHpYxtZVLiIifej7axaEbWwdoYuIxIguBbqZLTWzIjMrNrMvXuR1M7Pvd7y+18wW+i9VRETeyiUD3cwSgbXArcBMYI2Zzbxgs1uBqR1/7gb+x3OdIiJyCV05Ql8MFDvnjjjnWoD1wIoLtlkB/MKFvAJkmFnvllOJiEi3dCXQs4HSTo/LOp7r7jaY2d1mlmdmedXV1d2tVURE3kJXAv1iN9xzPdgG59yDzrlc51xuZmZmV+oTEZEu6kqglwHjOj3OAcp7sI2IiIRRVwJ9FzDVzCaZWQqwGth2wTbbgA90XO1yFVDnnKvwXKuIiLyFSy4scs61mdm9wA4gEfiJc+6Amd3T8foDwHbgXUAx0Ah86FLj5ufn15hZ+PpIRo+RQE2ki+hHtD/+TPviL2l/hEx4sxfMub+a6pY+ZGZ5zrncSNfRX2h//Jn2xV/S/rg0rRQVEYkRCnQRkRihQI+8ByNdQD+j/fFn2hd/SfvjEjSHLiISI3SELiISIxToIiIxQoHeh8xsnJk9Y2YHzeyAmX264/nhZvY7Mzvc8d9hka61r5hZopkVmNnjHY/jeV9kmNlGM3u943vk6njdH2b22Y6fkf1mts7MBsTrvugOBXrfagP+0Tl3OXAV8ImOVsRfBH7vnJsK/L7jcbz4NHCw0+N43hffA550zs0A5hHaL3G3P8wsG/gUkOucm01oQeNq4nBfdJcCvQ855yqcc7s7/l5P6Ac2m1D74Z93bPZzYGVECuxjZpYDLAN+3OnpeN0XQ4DrgIcAnHMtzrla4nR/EFrFPtDMkoA0Qr2h4nVfdJkCPULMbCKwAHgVGHW+903Hf7MiWFpf+i7weSDY6bl43ReTgWrgpx1TUD82s0HE4f5wzgWA/wRKgApCvaGeIg73RXcp0CPAzAYDjwKfcc6diXQ9kWBmy4Eq51x+pGvpJ5KAhcD/OOcWAA3E6ZRCx9z4CmASMBYYZGZ/G9mqooMCvY+ZWTKhMP+1c25Tx9OV5+/w1PHfqkjV14euAW4zs2OE7oJ1o5n9ivjcFxBqQV3mnHu14/FGQgEfj/vjZuCoc67aOdcKbALeRnzui25RoPchMzNCc6QHnXP/1emlbcDfdfz974CtfV1bX3POfck5l+Ocm0johNcfnHN/SxzuCwDn3Amg1Mymdzx1E/Aa8bk/SoCrzCyt42fmJkLnm+JxX3SLVor2ITO7FngB2Mef543/mdA8+iPAeELfzHc6505FpMgIMLMbgH9yzi03sxHE6b4ws/mEThCnAEcItaFOIA73h5ndD7yP0JVhBcBHgcHE4b7oDgW6iEiM0JSLiEiMUKCLiMQIBbqISIxQoIuIxAgFuohIjFCgi4jECAW6iEiM+P/pKXdZZnbv+QAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# x= [100/i for i in [1,2,3,4,5,6,7]]\n",
    "\n",
    "x=a[\"Group Size\"]\n",
    "y=a[\"Good Thresh\"]\n",
    "\n",
    "plt.plot(x,y)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-02T05:11:53.721595Z",
     "start_time": "2022-08-02T05:11:53.132579Z"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAD4CAYAAADlwTGnAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAVNElEQVR4nO3df4xdZX7f8fdn7boSdIl3yxC54C0QOU5RG8B7y4+2IepSVhhVuJsurfkHykb1Ol3QEqlRvUpUpaoqbbdJW1CQKWGNQoQgiwvJRKEQtGpaqYqpx7sOiwEvg7XUEwzMVl2QQF2X7rd/3OPmcrn2nPHM+AfP+yVd3Xue5zn3fs+5j/2Ze86dM6kqJEnt+djpLkCSdHoYAJLUKANAkhplAEhSowwASWrU6tNdwGKcf/75dfHFF5/uMiTprLJv377vV9XUePtZFQAXX3wxMzMzp7sMSTqrJHltUruHgCSpUQaAJDXKAJCkRhkAktQoA0CSGmUASFKjDABJalSvAEhyY5KDSWaT7JjQnyT3dv3PJ9nUtW9Msn/k9k6Su7u+f5Pk5W78k0nWLueGSZJObMEASLIKuA/YDFwG3JrksrFhm4EN3W0bsBOgqg5W1RVVdQXwaeA94MlunWeBv1pVPw18F/jKkrdGktRbn08AVwGzVXWoqo4CjwFbxsZsAR6uoT3A2iTrxsZcD7xaVa8BVNUfVtX7Xd8e4KKT3gpJ0qL1CYALgcMjy3Nd22LHbAUePc5rfAH4T5M6kmxLMpNkZn5+vke5kqQ++gRAJrSN/x3JE45Jsga4GXj8Q0+e/DLwPvDIpBevqgeqalBVg6mpD13LSJJ0kvpcDG4OWD+yfBHw+iLHbAa+VVVvjq6U5Hbg7wLXl3+cWJJOqT6fAPYCG5Jc0v0kvxWYHhszDdzWfRvoGuDtqjoy0n8rY4d/ktwI/DPg5qp676S3QJJ0Uhb8BFBV7ye5E3gGWAXsqqoDSbZ3/fcDTwE3AbMMv+lzx7H1k5wD3AB8ceypfwP488CzSQD2VNX2JW+RJKmXnE1HXgaDQfn3ACRpcZLsq6rBeLu/CSxJjTIAJKlRBoAkNcoAkKRGGQCS1CgDQJIaZQBIUqMMAElqlAEgSY0yACSpUQaAJDXKAJCkRhkAktQoA0CSGmUASFKjDABJapQBIEmNMgAkqVEGgCQ1ygCQpEb1CoAkNyY5mGQ2yY4J/Ulyb9f/fJJNXfvGJPtHbu8kubvruyXJgSQ/SvKhP1YsSVpZqxcakGQVcB9wAzAH7E0yXVUvjgzbDGzoblcDO4Grq+ogcMXI8/wp8GS3zgvAzwH/YVm2RJK0KH0+AVwFzFbVoao6CjwGbBkbswV4uIb2AGuTrBsbcz3walW9BlBVL3UBIUk6DfoEwIXA4ZHlua5tsWO2Ao8utsAk25LMJJmZn59f7OqSpOPoEwCZ0FaLGZNkDXAz8Hj/0ronqXqgqgZVNZiamlrs6pKk4+gTAHPA+pHli4DXFzlmM/CtqnrzZIqUJC2/PgGwF9iQ5JLuJ/mtwPTYmGngtu7bQNcAb1fVkZH+WzmJwz+SpJWzYABU1fvAncAzwEvAN6rqQJLtSbZ3w54CDgGzwG8C/+TY+knOYfgNoidGnzfJ55LMAdcCf5DkmWXYHklST6kaP5x/5hoMBjUzM3O6y5Cks0qSfVX1od+38jeBJalRBoAkNcoAkKRGGQCS1CgDQJIaZQBIUqMMAElqlAEgSY0yACSpUQaAJDXKAJCkRhkAktQoA0CSGmUASFKjDABJapQBIEmNMgAkqVEGgCQ1ygCQpEYZAJLUqF4BkOTGJAeTzCbZMaE/Se7t+p9Psqlr35hk/8jtnSR3d32fTPJskle6+08s65ZJkk5owQBIsgq4D9gMXAbcmuSysWGbgQ3dbRuwE6CqDlbVFVV1BfBp4D3gyW6dHcA3q2oD8M1uWZJ0ivT5BHAVMFtVh6rqKPAYsGVszBbg4RraA6xNsm5szPXAq1X12sg6v9U9/i3g753MBkiSTk6fALgQODyyPNe1LXbMVuDRkeUfr6ojAN39BZNePMm2JDNJZubn53uUK0nqo08AZEJbLWZMkjXAzcDj/UvrnqTqgaoaVNVgampqsatLko6jTwDMAetHli8CXl/kmM3At6rqzZG2N48dJuru3+pbtCRp6foEwF5gQ5JLup/ktwLTY2Omgdu6bwNdA7x97PBO51Y+ePjn2Dq3d49vB35v0dVLkk7a6oUGVNX7Se4EngFWAbuq6kCS7V3//cBTwE3ALMNv+txxbP0k5wA3AF8ce+qvAt9I8vPA/wBuWfrmSJL6StX44fwz12AwqJmZmdNdhiSdVZLsq6rBeLu/CSxJjTIAJKlRBoAkNcoAkKRGGQCS1CgDQJIaZQBIUqMMAElqlAEgSY0yACSpUQaAJDXKAJCkRhkAktQoA0CSGmUASFKjDABJapQBIEmNMgAkqVEGgCQ1qlcAJLkxycEks0l2TOhPknu7/ueTbBrpW5tkd5KXk7yU5Nqu/fIkf5zkO0l+P8l5y7dZkqSFLBgASVYB9wGbgcuAW5NcNjZsM7Chu20Ddo703QM8XVU/BVwOvNS1PwjsqKq/BjwJ/NIStkOStEh9PgFcBcxW1aGqOgo8BmwZG7MFeLiG9gBrk6zrfqq/Dvg6QFUdraofdOtsBP5r9/hZ4O8vbVMkSYvRJwAuBA6PLM91bX3GXArMAw8l+XaSB5Oc2415Abi5e3wLsH6RtUuSlqBPAGRCW/UcsxrYBOysqiuBd4Fj5xC+AHwpyT7g48DRiS+ebEsyk2Rmfn6+R7mSpD76BMAcH/zp/CLg9Z5j5oC5qnqua9/NMBCoqper6rNV9WngUeDVSS9eVQ9U1aCqBlNTUz3KlST10ScA9gIbklySZA2wFZgeGzMN3NZ9G+ga4O2qOlJVbwCHk2zsxl0PvAiQ5ILu/mPArwD3L31zJEl9rV5oQFW9n+RO4BlgFbCrqg4k2d713w88BdwEzALvAXeMPMVdwCNdeBwa6bs1yZe6x08ADy3D9kiSekrV+OH8M9dgMKiZmZnTXYYknVWS7KuqwXi7vwksSY0yACSpUQaAJDXKAJCkRhkAktQoA0CSGmUASFKjDABJapQBIEmNMgAkqVEGgCQ1ygCQpEYZAJLUKANAkhplAEhSowwASWqUASBJjTIAJKlRBoAkNcoAkKRG9QqAJDcmOZhkNsmOCf1Jcm/X/3ySTSN9a5PsTvJykpeSXNu1X5FkT5L9SWaSXLV8myVJWsiCAZBkFXAfsBm4DLg1yWVjwzYDG7rbNmDnSN89wNNV9VPA5cBLXfvXgH9RVVcA/7xbliSdIn0+AVwFzFbVoao6CjwGbBkbswV4uIb2AGuTrEtyHnAd8HWAqjpaVT/o1ingvO7xjwGvL21TJEmLsbrHmAuBwyPLc8DVPcZcCLwPzAMPJbkc2Ad8uareBe4GnknyawyD6G9MevEk2xh+quBTn/pUj3IlSX30+QSQCW3Vc8xqYBOws6quBN4Fjp1D+AXgF6tqPfCLdJ8SPvQkVQ9U1aCqBlNTUz3KlST10ScA5oD1I8sX8eHDNccbMwfMVdVzXftuhoEAcDvwRPf4cYaHmiRJp0ifANgLbEhySZI1wFZgemzMNHBb922ga4C3q+pIVb0BHE6ysRt3PfBi9/h14Ge7x58BXlnKhkiSFmfBcwBV9X6SO4FngFXArqo6kGR7138/8BRwEzALvAfcMfIUdwGPdOFxaKTvHwP3JFkN/G+64/ySpFMjVeOH889cg8GgZmZmTncZknRWSbKvqgbj7f4msCQ1ygCQpEYZAJLUKANAkhplAEhSowwASWqUASBJjTIAJKlRBoAkNcoAkKRGGQCS1CgDQJIaZQBIUqMMAElqlAEgSY0yACSpUQaAJDXKAJCkRhkAktQoA0CSGtUrAJLcmORgktkkOyb0J8m9Xf/zSTaN9K1NsjvJy0leSnJt1/47SfZ3t+8l2b9sWyVJWtDqhQYkWQXcB9wAzAF7k0xX1YsjwzYDG7rb1cDO7h7gHuDpqvp8kjXAOQBV9Q9HXuPXgbeXvjmSpL4WDADgKmC2qg4BJHkM2AKMBsAW4OGqKmBP91P/OuBd4DrgHwFU1VHg6OiTJwnwD4DPLG1TJEmL0ecQ0IXA4ZHlua6tz5hLgXngoSTfTvJgknPH1v0Z4M2qemXSiyfZlmQmycz8/HyPciVJffQJgExoq55jVgObgJ1VdSXDTwTj5xBuBR493otX1QNVNaiqwdTUVI9yJUl99AmAOWD9yPJFwOs9x8wBc1X1XNe+m2EgAJBkNfBzwO8srmxJ0lL1CYC9wIYkl3QncbcC02NjpoHbum8DXQO8XVVHquoN4HCSjd246/nguYO/A7xcVXNL2wxJ0mIteBK4qt5PcifwDLAK2FVVB5Js7/rvB54CbgJmgfeAO0ae4i7gkS48Do31beUEh38kSSsnwy/unB0Gg0HNzMyc7jIk6aySZF9VDcbb/U1gSWqUASBJjTIAJKlRBoAkNcoAkKRGGQCS1CgDQJIaZQBIUqMMAElqlAEgSY0yACSpUQaAJDXKAJCkRhkAktQoA0CSGmUASFKjDABJapQBIEmNMgAkqVEGgCQ1qlcAJLkxycEks0l2TOhPknu7/ueTbBrpW5tkd5KXk7yU5NqRvru65z2Q5GvLs0mSpD5WLzQgySrgPuAGYA7Ym2S6ql4cGbYZ2NDdrgZ2dvcA9wBPV9Xnk6wBzume928DW4CfrqofJrlgmbZJktRDn08AVwGzVXWoqo4CjzH8j3vUFuDhGtoDrE2yLsl5wHXA1wGq6mhV/aBb5xeAr1bVD7u+t5a+OZKkvvoEwIXA4ZHlua6tz5hLgXngoSTfTvJgknO7MT8J/EyS55L8lyR/fdKLJ9mWZCbJzPz8fI9yJUl99AmATGirnmNWA5uAnVV1JfAucOwcwmrgE8A1wC8B30jyoeepqgeqalBVg6mpqR7lSpL66BMAc8D6keWLgNd7jpkD5qrqua59N8NAOLbOE91ho/8O/Ag4f3HlS5JOVp8A2AtsSHJJdxJ3KzA9NmYauK37NtA1wNtVdaSq3gAOJ9nYjbseOHby+HeBzwAk+UlgDfD9JW2NJKm3Bb8FVFXvJ7kTeAZYBeyqqgNJtnf99wNPATcBs8B7wB0jT3EX8EgXHodG+nYBu5K8ABwFbq+q8UNLkqQVkrPp/9zBYFAzMzOnuwxJOqsk2VdVg/F2fxNYkhplAEhSowwASWqUASBJjTIAJKlRBoAkNcoAkKRGGQCS1CgDQJIaZQBIUqMMAElqlAEgSY0yACSpUQaAJDXKAJCkRhkAktQoA0CSGmUASFKjDABJapQBIEmN6hUASW5McjDJbJIdE/qT5N6u//kkm0b61ibZneTlJC8lubZr/9Ukf5pkf3e7afk2S5K0kNULDUiyCrgPuAGYA/Ymma6qF0eGbQY2dLergZ3dPcA9wNNV9fkka4BzRtb7d1X1a0vfDEnSYvX5BHAVMFtVh6rqKPAYsGVszBbg4RraA6xNsi7JecB1wNcBqupoVf1g+cqXJJ2sPgFwIXB4ZHmua+sz5lJgHngoybeTPJjk3JFxd3aHjHYl+cSkF0+yLclMkpn5+fke5UqS+ugTAJnQVj3HrAY2ATur6krgXeDYOYSdwE8AVwBHgF+f9OJV9UBVDapqMDU11aNcSVIffQJgDlg/snwR8HrPMXPAXFU917XvZhgIVNWbVfV/q+pHwG8yPNQkSTpF+gTAXmBDkku6k7hbgemxMdPAbd23ga4B3q6qI1X1BnA4ycZu3PXAiwBJ1o2s/znghaVsiCRpcVI1fjRnwqDhVzT/PbAK2FVV/yrJdoCquj9JgN8AbgTeA+6oqplu3SuAB4E1wKGu738l+W2Gh38K+B7wxao6skAd88Bri97KofOB75/kuivJuhbHuhbHuhbnTK0LllbbX66qDx1D7xUAHwVJZqpqcLrrGGddi2Ndi2Ndi3Om1gUrU5u/CSxJjTIAJKlRLQXAA6e7gOOwrsWxrsWxrsU5U+uCFaitmXMAkqQPaukTgCRphAEgSY06KwOgu3bQW0leGGm7PMkfJ/lOkt/vLkQ3ad2Jl7ZO8skkzyZ5pbufeG2ilagryfok/7m7XPaBJF8e6VvyZbOXuL++143Zn2RmpP107q+NI/tjf5J3ktzd9S3H/pr4fvTd5pWaY0upayXn2DLsrxWZY0vcXys2x05Q1y3d8o+SHPfrnss6v6rqrLsxvMLoJuCFkba9wM92j78A/MsJ660CXmV4kbo1wJ8Al3V9XwN2dI93AP/6FNa1DtjUPf448N2Run4V+KenY391fd8Dzp/Qftr214T39A2Gv+iyXPtr4vvRZ5tXco4tsa4Vm2NLqWsl59hS61qpOXaCuv4KsBH4I2BwglqWbX6d9D+S030DLuaD/3G8w5+d1F4PvDhhnWuBZ0aWvwJ8pXt8EFg38gYdPFV1TXiO3wNuWI7JttS6TvCP84zYX8Bngf82srws+2vS+9Fnm0/FHDuZuk7FHDvZulZ6ji11f630HBt9L7rlP+L4AbCs8+usPAR0HC8AN3ePb+GDF6c75kSXtv7x6i5F0d1fcArr+v+SXAxcCTw30rzgZbNXsK4C/jDJviTbRtrPiP3F8NpUj461Ldv+Gns/+mzzKZljJ1HX8dY9Zln22UnWteJzbCn7ixWcY8d5L05kWefXRykAvgB8Kck+hh+rjk4Y0+fS1sutT10AJPkLwH8E7q6qd7rmXpfNXsG6/mZVbWL4V9++lOS6ZXr9pdZFhhcnvBl4fKR52fbXcd6PBVeb0Lasc+wk6zrRusuyz5ZQ14rOsSXurxWbY2fC/PrIBEBVvVxVn62qTzNM61cnDDvRpa3fTHeF0u7+rVNYF0n+HMPJ8EhVPTGy/opcNrtvXVX1enf/FvDkyOuf1v3V2Qx8q6reHFl/WfbXcd6PPtu8onNsCXWt6BxbSl0rOceWUldnRebY8d6LHpZ1fn1kAiDJBd39x4BfAe6fMOxEl7aeBm7vHt/O8LjcKakrSRj+2cyXqurfjvWtyGWze9Z1bpKPH3vM8Fjosdc/bftrxK2MfTRfjv11gvejzzav2BxbSl0rOceWWNeKzbElvo/HLPscO9F70cPyzq/lOpFxKm8M35AjwP9hmIg/D3yZ4dn07wJf5c9OJP4l4KmRdW/qxrwK/PJI+18Evgm80t1/8lTVBfwthh/jngf2d7ebur7fBr7T9U3TneQ5RXVdyvBbBn8CHDhT9le3fA7wP4EfG3vO5dhfE9+P423zqZpjS6lrJefYEutasTm2DO/jisyxE9T1OYb/Dn4IvEl3sncl55eXgpCkRn1kDgFJkhbHAJCkRhkAktQoA0CSGmUASFKjDABJapQBIEmN+n+vhFVyKRESPgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "x=a[\"Group Size\"][:]\n",
    "y=a[\"Bad Thresh\"][:]\n",
    "\n",
    "plt.plot(x,y)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-17T01:05:21.705716Z",
     "start_time": "2022-06-17T01:05:21.593982Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD4CAYAAAD8Zh1EAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAhj0lEQVR4nO3deXxU1f3/8dcnk4SQhbAlLGEJOyICyiIquIugtbRftXVfAWlrq9VqtbXa5ftrrbWLtVa+FBdcWmpdClUqWnerQkDZFwlhSUggYQsJkHXO748JGjHAkMzNTG7ez8cjj8yde3POx/tI3hzP3HuPOecQEZGWLy7aBYiISGQo0EVEfEKBLiLiEwp0ERGfUKCLiPhEfLQ67ty5s8vOzo5W9yIiLdKSJUt2OOcyGtoXtUDPzs5m8eLF0epeRKRFMrPNh9unKRcREZ9QoIuI+IQCXUTEJxToIiI+oUAXEfGJowa6mT1uZsVmtvIw+83M/mhmuWa23MxOinyZIiJyNOGM0J8EJh5h/yRgQN3XNODRppclIiLH6qjXoTvn3jWz7CMcMhl4yoWew/uRmbU3s27OuaJIFVnfum1lvLK80Iumj1m/zFQmj8hq9M+v3FrKttIKzh3SpdFtLM3fQ0V1LWP7dmp0GwvzdpKcGM8JPdIb3YaIRF8kbizKAvLrbRfUvfelQDezaYRG8fTq1atRneUWl/PwW7mN+tlIcg4SAtboQJ+7dCt3PL+czimJjQ704rIKrn18Eb07JTPv5nGNaqOsopqpTy1mVHZHHr9udKPaEJHYEIlAtwbea3DVDOfcTGAmwKhRoxq1ssaFw7px4bALG/OjEfXb19bxSCP+YXHO8fCbufzu9U8xg9pGLjDinOPHL62k9EA1tcHGL1Lyt0Vb2FtR06Q2RCQ2RCLQC4Ce9bZ7ALExJxJjKmtqufuFFbz4yVb+58TQyP6/G3Y0qq15ywp5ffV2EgIN/Xsafj2Pvb+x0T8vIrElEpctzgOuqbvaZSxQ6tX8eUu2e18VV89axIufbOX28wby228MJzG+cae/uKyC++at4sRe7RnXv3Oja5r7SSHb91Y2ug4RiS1HHaGb2d+AM4HOZlYA3AckADjnZgDzgQuAXGA/cL1XxbZUeSXl3PBkDoWlFTx02YgmfZDqnOOel1ayv6qW31wynPv/vaZR7QSDjhnvbuD47u2Ij2v8KF9EYkc4V7lcfpT9DvhOxCrymY/ydjL9mSXEmfG3qSczsnfHJrU3b1khr63ezt2TBtM/M7XR7by+Zjt5Jfv44+UnatpFxCf0/9oeemFJAVc/tpBOKYn889unNTnMS8oquW/eKkb0bM+U8X0b3Y5zjkff3kDPjm25YGjXJtUkIrEjas9D97Ng0PH7/3zKw2/mclr/Tvz5ypGkt01oUpvOOe755wr2V9Xy4KXDCDRhmmThxl0szd/DL742lPiA/k0X8QsFeoRVVNfyg38s4+XlRXxzVE/+9+tDSYhAaP5reRELVm3nrkmD6Z+Z1qS2Zryzgc6piVw6skeT6xKR2KFAj6Ad5ZVMe2oxH2/Zw92TBjPt9L6YNf0Dx5KySu6bu5LhPdszZVyfJrW1pmgvb68r4QcTBpKUEGhybSISOxToEbJ+exk3zM6hpKySGVedxMSh3SLSrnOOn/xzJfsqa3nwkmFNniKZ8c4GUhIDXD02OyL1iUjsUKBHwPvrd/CtZ5eQlBDg79NOYXjP9hFr+5UVRby6ahs/nDiYAV2aNtWSv2s/Ly8v4obTsklPbtqcvojEHgV6E81ZtIV7/rmS/pmpPHbdaLLat41Y2zvKK7l37iqG90hn6vimTbUAzHovjziDG8c1/goZEYldCvQmCDq468UVnDEwgz9dcSJpSZEd9d47dyXlFTU8eOnwJk+17Cyv5O+L8/naiCy6pidFqEIRiSUK9EaKjwsF7NVje3PfRUMifvnfK8uLmL9iG3dOHNTkqRaA2R9soqI6yE1naHQu4lcK9Ea64uReDOuRzpmDMiJyJUt9O8or+cnclQzrkc60JtxAdNC+yhpmf7iZCUO6NPmSRxGJXQr0RspIa8NZgzM9afu+uasiNtUCMCcnn9ID1Uw/s18EqhORWKXbBGPMK8uLeGVFEbecO4CBEZhqqa4N8th7eYzp05GTenWIQIUiEqsU6DFkZ3kl985dyQlZ6dx0emTmuuctLaSwtIJvnaHRuYjfKdBjyL3zVlEWwamWYNAx450NDO6axpmDMiJQoYjEMgV6jJi/oohXloemWgZ1jcwHl2+uLWZ9cTnTz+gX8Q9uRST2KNBjwM7ySn7yz8hOtUDoNv+s9m35yrDIPIZARGKbAj0G3DdvFXsrqiM21QKQs2kXizfvZur4PnpErkgrob/0KCsuq+Tl5UXcck7kploAZry9gQ7JCXxjdM+jHywivqBAjzLnYGhWO26K4FUo67aV8cbaYq47tQ/JibrVQKS1UKBHUUIgjoSA8eClwyOyCMZB//fuBtomBLjmlN4Ra1NEYp+Gb1H07bP68fWTshjctV3E2ty65wDzlhZy9Sm96ZCSGLF2RST2KdCjqFt6W7qlR+5xuxB6RC7QpEWkRaRl0pSLj+zeV8WcRfl8dXj3iD6XXURaBgW6jzz14WYOVNdG9ANWEWk5FOg+caCqltkfbuKcwZkRvfxRRFoOBbpPPLc4n137qvSIXJFWTIHuAzW1jpnv5jGydwdGZ3eMdjkiEiW6ysUH1m0vA+BnXz0+ypWISDRphO4TAzJTOdujFZREpGVQoPvETWf0Iy5Oj8gVac0U6C1ccmI8We3b8tXh3aNdiohEWVhz6GY2EXgICACznHP3H7I/HXgG6FXX5oPOuSciXKs04L6LhlBd60iM17/NIq3dUQPdzALAI8B5QAGQY2bznHOr6x32HWC1c+4iM8sA1pnZs865Kk+qls90Sm0T7RJEJEaEM6wbA+Q65/LqAnoOMPmQYxyQZqF1zlKBXUBNRCsVEZEjCifQs4D8etsFde/V9yfgOKAQWAHc4pwLHtqQmU0zs8VmtrikpKSRJYuISEPCCfSGLp1wh2yfDywFugMjgD+Z2ZeeCeucm+mcG+WcG5WRoVXoRUQiKZxALwDqr2PWg9BIvL7rgRddSC6wERgcmRJFRCQc4QR6DjDAzPqYWSJwGTDvkGO2AOcAmFkXYBCQF8lCRUTkyI56lYtzrsbMbgYWELps8XHn3Cozm163fwbwC+BJM1tBaIrmh865HR7WLSIihwjrOnTn3Hxg/iHvzaj3uhCYENnSRETkWOhuFBERn9DTFgWAA9W1FO45QEIgjsRAHPEBIyEQR0LACN1eICKxToEutAnEsWjjLk69/80G9yfUhXt8nJEYH1cX9HGfvX/w9bTT+zFxaNdmrl5EDlKgC7+6+ASWbtlDdW2Q6qCjuiYYel0bpKrWUVN7cNtRVRv8fH+9Y/+7YSf/WbNdgS4SRQp0oV9GKv0yUpvUxmmHGd2LSPPRh6IiIj6hQBcR8QkFuoiITyjQRUR8QoEuIuITCnQREZ9QoIuI+IQCXUTEJxToIiI+oUAXEfEJBbqIiE8o0EVEfEKBLiLiEwp0ERGfUKCLiPiEAl1ExCcU6CIiPqFAFxHxCQW6xIT9VTVc8NB7vLCkINqliLRYCnSJCY+/v5HVRXtZX1we7VJEWiwFukTdzvJKZryTF+0yRFo8BbpE3cNv5rK/qgazaFci0rIp0CWqtuzcz7MLN/PN0T1JCOjXUaQp9BckUfXga+sIxBm3njsw2qWItHgKdImaFQWlzFtWyI3j+tClXVK0yxFp8cIKdDObaGbrzCzXzO46zDFnmtlSM1tlZu9EtkzxG+cc97+6hg7JCdx0Rr9olyPiC/FHO8DMAsAjwHlAAZBjZvOcc6vrHdMe+DMw0Tm3xcwyPapXfOLd9Tv4b+5O7v3KENolJUS7HBFfCGeEPgbIdc7lOeeqgDnA5EOOuQJ40Tm3BcA5VxzZMsVPgkHH/f9eS8+ObblybK9olyPiG+EEehaQX2+7oO69+gYCHczsbTNbYmbXNNSQmU0zs8VmtrikpKRxFUuLN3fZVtYU7eUHEwbRJj4Q7XJEfCOcQG/o6mB3yHY8MBK4EDgf+ImZfemyBefcTOfcKOfcqIyMjGMuVlq+iupaHlzwKUOz2nHRsO7RLkfEV446h05oRN6z3nYPoLCBY3Y45/YB+8zsXWA48GlEqhTfeOajzWzdc4BfXzyMuDjdSSQSSeGM0HOAAWbWx8wSgcuAeYccMxcYb2bxZpYMnAysiWyp0tKVHqjmT2/lMn5AZ8YN6BztckR856iB7pyrAW4GFhAK6eecc6vMbLqZTa87Zg3wKrAcWATMcs6t9K5saYkefXsDpQequWvSYE/aDwYdj72/kcI9BzxpXyTWhTPlgnNuPjD/kPdmHLL9G+A3kStN/KRwzwGe+O9GvjYii+O7p3vSx7+WF/KLl0NX0944ro8nfYjEMt0pKs3i969/inNw23ne3OJfUV3LA6+uA0I3LYm0Rgp08dy6bWW88HEB15zSm54dkz3p46kPN7FVUy3SyinQxXMPvLqWlDbxfOes/p60v2d/FX96M5eRvTt40r5IS6FAF08tzNvJG2uL+daZ/eiQkuhJHw+/mUt5ZQ13e/Rhq0hLoUAXzzjn+NW/19K1XRI3nObNh5Rbdu7nqQ83cenIngzsmuZJHyIthQJdPPPvldtYmr+H284bSFKCN7f4P7BgLfFxcdw2Qc9TF1Ggiyeqa4P8ZsE6BnZJ5eKRPTzpY2n+Hl5eXsTU8Xqeuggo0MUjcxZtYeOOffxw4mACHtzi75zjl6+soXNqItP0PHURQIEuHthXWcNDb6xnTJ+OnD3Ym0fjv756O4s27eLWcweS2ias++NEfE+BLhH3l/fy2FFexV2TBmMW+dF5dW2Q+19dS7+MFC4b3fPoPyDSSijQJaJKyiqZ+W4ek4Z25aRe3lwXPicnn7ySfdw16TjiA/oVFjlIfw0SUX98Yz2VNUHuOH+QJ+2XV9bw0H8+ZUyfjpx7nFY6FKlPk48SMZt27GNp/h4uH9OTvhmpnvTxf+9sYEd5FY9de5wn0zkiLZlG6BIxizfvJjE+jlvO8eaa8G2lFfzlvTwuGt6d4T3be9KHSEumQJeImjq+LxlpbTxp+3evryMYhDs9ms4RaekU6BIxnVMTmXp6X0/aXrttL/9Y4u0TG0VaOs2hS0TcdEZfenVM9uya8F/NX0tam3huPtubJzaK+IFG6BIR15ySzZmDvLnq5P31O3jn0xK+e/YA2id788TGg0oPVHP3i8v1bHVpkRToEtOCQccv56+hR4e2XHNqb8/7++m8VfxtUT6LN+3yvC+RSFOgS0x76ZOtrC7ayx3nD6JNvDdPbDzo1ZVFvPTJVk/7EPGSAl1iVkV1LQ++to5hPdK5aFh3T/sqKavkRy+tJNOjK3REmoMCXWLWY+9vpKi0gh9dcBxxHjyx8SDnHD96aQXllTX8fPJQz/oR8ZoCXWLSzvJKHn17A+cel8nYvp087ev5JQW8vno7d0wYxMAu3tzhKtIcFOgSk/74xnoOVNdyl8frhG7dc4Cf/2s1Y7I7csM4b5bJE2kuCnSJOXkl5Ty7cAvfHN2T/pnerRMaDDru+Mcygs7x4KXDPVmIQ6Q5KdAl5jzw6joS4+O49dwBnvYz+8NNfLBhJ/d8ZQi9OunuU2n5dKeoxJSPN+9m0aZdfP/cgWSmebdO6IaScu7/91rOHJShRTLENzRCl5iyaNMuMtPaMPV07+aza2qD3PbcMtomBnjg4mF6DK/4hkboEnNunzCQ5ETvfjVnvLOBZfl7ePjyE8ls593/BYg0N43QJWYYMKhLGpeM9G4KZOXWUv7wn/V8ZVg3Lhru7c1KIs1NI3SJGfdceBwje3f07GqTyppabn9uGR1TEvmFbiASHwprhG5mE81snZnlmtldRzhutJnVmtklkStRWourT8lmSPd2nrX/u9c/Zd32Mn598TA6pHj71EaRaDhqoJtZAHgEmAQMAS43syGHOe7XwIJIFynSVDmbdjHz3TwuH9OTswZrcWnxp3BG6GOAXOdcnnOuCpgDTG7guO8CLwDFEaxPpMn2VdZw+3PL6NGhLT++8EtjERHfCCfQs4D8etsFde99xsyygK8DM47UkJlNM7PFZra4pKTkWGsVaZRfzl9D/u79/PbSEZ6tqHSoA1W1zdKPSH3hBHpDn1C5Q7b/APzQOXfE32Ln3Ezn3Cjn3KiMjIwwSxRpvHc+LeHZhVuYMq4PY/p0bJY+5yzawtCfLmDttr3N0p/IQeEMVwqA+teR9QAKDzlmFDCn7gaNzsAFZlbjnPtnJIoUaYzS/dXc+fwyBmSmcvuEQc3S5+urt/Ojl1YQdLCrvKpZ+hQ5KJxAzwEGmFkfYCtwGXBF/QOcc5/d1mdmTwIvK8wl2u6dt5Kd5VU8du1okhK8Xe0IYMnmXdz8149Jb5vA7v3VnvcncqijTrk452qAmwldvbIGeM45t8rMppvZdK8LFGmMV5YXMXdpId89ewBDs9I97y+3uIwbZy+me/u2/PLrJ3jen0hDwvqEyDk3H5h/yHsNfgDqnLuu6WWJNF5JeSWPvJXL8B7pfPusfp73t620gmsfzyE+Lo7Z14+hqPSA532KNES3/ovvPPH+JvZX1fLbb4wgIeDtr3jpgWque2IRpQeqefL60XoMr0SVAl18p6o2yJ0TB9M/09vl5Cqqa5n21GI2lJQz46qRzTK1I3IkepaL+Eag7jG4Y/t25PpTsz3tqzbouO25pSzcuIuHLhvBuAGdPe1PJBwKdPGNlDbx/PnKkzi5T0fiPFxOzjnHz/+1ivkrtnHPhccxeUTW0X9IpBko0MVXLjihm+d9PPrOBmZ/uJmp4/swZXxfz/sTCZfm0EWOwfNLCnjg1XVMHtGduycdF+1yRL5AgS4SprfWFfPDF5Yzrn9nfnPJcE+ndUQaQ4EuEoal+Xv49jMfM7hrGo9edRKJ8d7+6bz4cQHjH3iT3fv0+AAJn+bQRY5i44593PBkDp3TEnni+tGkJSV42t+/lhXyg38sI+hCN0lpMQ4Jl0boIkdQXFbBNY8vBOCpG04mM83bRaUXrNrGrX9fSkozPeZX/EWBLnIY5ZU1XP9EDjvKqnjiutH06ZziaX9vrSvm5r9+zAlZ6fxEC3FIIyjQRRpQVRNk+tNLWLutjD9fdRLDe7b3tL8Pcncw/eklDOySxuwbxmiELo2i3xqRBvz2tU/ZtreCBy8dzlmDvF2DNGfTLm6cvZjsTik8fePJpLf1do5e/EsjdJEGbNtbwZ0TB3HJyB6e9rM0fw/XP5FDt/QknplyMh31Aag0gQJdpJ6Da45ed2o23zrD20fvrios5ZrHFtIhJYFnp55MRlobT/sT/9OUi0g9me2S+PDus+naLom6JRU98en2Mq5+bBGpbeL565SxdEtv61lf0noo0EUO4XW4btyxjytnLSQ+znh26lh6dtQz1CUyNOUi0ozyd+3nir98RDDo+OvUkz2/FFJaF43QRZpJUekBrpj1Efuravnb1LH0z0yLdkniMxqhizSD4rIKrvzLQvbsq+bpG8cwpHu7aJckPqQRuojHdu2r4qpZC9m2t4KnbxzDsB7to12S+JRG6CIeKt1fzVWzFrJ5534eu3Y0I3t3jHZJ4mMKdBGPlFXWcM0Ti8gtLmfmNaM4pV+nZu0/t7icZxdubtY+Jbo05SLikR+/tJI9+6t49KqRnDEwo9n6dc7xzEeb+X/z11BRHeTCE7rRPll3oLYGCnQRj+zaV8nDl5/EeUO6NFufO8orufP55by5tpgOyQlUVAcJumbrXqJMgS4SYZntkkhrE8/PJh/PhcO8X7T6oLfWFnPH88vYW1HDfReFHr/7s3+tbrb+JfoU6CIR1qdzCsvum9Bsa45WVNfyy/lreOrDzQzumsazU8YyqGsasz/Y1Cz9S+xQoIt4oLnCfFVhKbfMWUpucTk3juvDHecPIikh0Cx9S+xRoIu0QMGgY9b7efxmwTo6JCfy9I1jGD+g+T54ldikQBdpYYpKD3D7c8v4YMNOzj++C/f/zzAtJC1AmNehm9lEM1tnZrlmdlcD+680s+V1Xx+Y2fDIlyoirywvYuIf3mNp/h5+ffEJzLhqpMJcPnPUEbqZBYBHgPOAAiDHzOY55+p/fL4ROMM5t9vMJgEzgZO9KFikNSqvrOGn81bx/JIChvdszx++OUJPapQvCWfKZQyQ65zLAzCzOcBk4LNAd859UO/4jwBv1+0SaUWWbN7N9/++lILd+/nu2f353jkDSAjoJm/5snACPQvIr7ddwJFH3zcC/25KUSISMuPtDcxdVki39CT+ftMpjM7Ws2Dk8MIJ9Iauv2rw3jMzO4tQoI87zP5pwDSAXr16hVmiSOv14idb+fqJWfxs8vG0S0qIdjkS48IJ9AKgZ73tHkDhoQeZ2TBgFjDJObezoYacczMJza8zatQo3ZAschjZnZPp2zmFW84dwOQRWdEuR1qIcAI9BxhgZn2ArcBlwBX1DzCzXsCLwNXOuU8jXqVIK3N893Te/MGZ0S5DWpijBrpzrsbMbgYWAAHgcefcKjObXrd/BnAv0An4c91K6TXOuVHelS0iIocK68Yi59x8YP4h782o93oKMCWypYmIyLHQtU8iIj6hQBcR8QkFuoiITyjQRUR8QoEu4nPbSiuiXYI0EwW6iE+N7duJtKR4Ln70A576cBNBLS7qewp0EZ8a1DWN175/OmP6dOTeuau4ctZC8nftj3ZZ4iEFuoiPdUtvy5PXj+bXF5/Aiq2lTPzDuzy7cDPOabTuRwp0EZ8zM745uhcLvn86J/bqwI9fWsk1jy+icM+BaJcmEaZAF2klstq35ekbx/C/XxvKks27Of/37/JcTr5G6z6iQBdpRcyMq8b2ZsGtp3N8VjvufGE5NzyZoythfEKBLtIK9eyYzF+njOWnFw3hw7ydTPj9O7z4cYFG6y2cAl2klYqLM647rQ+v3nI6A7ukcdtzy5j61BKKyzRab6kU6CKtXHbnFP5+0yncc+FxvLe+hAm/f5e5S7dqtN4CKdBFhECcMWV8X1753niyO6Vwy5ylfOuZj9lRXhnt0uQYKNBF5DP9M1N5fvop3DVpMG+uLWbC79/l9dXbo12WhEmBLiJfEB+IY/oZ/Xj5e+PISG3DrXM+iXZJEiYFuog0aGCXNCYc34UD1bXRLkXCpEAXEfEJBbqIiE+EtUi0iEg4qmuDbN65j9zifWwoKSe3uJwNJeUM7prGA5cMj3Z5vqdAF5FjVl5Zw4bizwM7t7ic3JJytuzcT0295653T0+isibI9r26Wak5KNBF5Igc8PSHm9hQsi8U3MXlbKsX0PFxRnbnFAZkpjJpaFf6Z6bSLyP0ldImnrteWM5b64qj9x/QiijQReSw2iYGcA5+MncVKYkB+mWmcmq/TvSrC+3+man07pRMQkAfx8UCBbqIHNZ1p2YzslcHenVKpmu7JMws2iXJESjQReSwkhPjOblvpya341xo3r28ooayimr2VtRQXhl6XVbx+ftllTWU1b0u/+x16GvS0K784mtDI/Bf5V8KdBHxlJlRXFbJ0PsWHOU4SE2MJy0pntSkeNKSEuiYkkivjsks2bybj7fsbqaKWy4Fuoh46rpTs8lITSSlTSikDwZ2u7rQTm0TCvGUxHji4hqe0pkyO4ciLcJxVAp0EfHUoK5pDOo6qMntlFXU8NbaYuIDRiDOiI+LIxBnJBxhOz5gxMd9vp2UEOfrzwEU6CIS89olJbBl136ufzKnSe1cd2o2P/3q8RGqKvYo0EUk5v3yf07g+tP6UBMMUht01AQdNbXu8Nu1ofdqg8HP9j3+342s2FrK6sK9BOKMQBzEWWj0fvB7fJwRF2cErO573etA3MHjiOkRfliBbmYTgYeAADDLOXf/Ifutbv8FwH7gOufcxxGuVURaqaSEACf0SG9SG2+s3c5Hebu44I/vNamduycN5qYz+jWpDa8cNdDNLAA8ApwHFAA5ZjbPObe63mGTgAF1XycDj9Z9FxGJCb/7xghWFe6lNugIOkdtsN6XcwTrvtd/P3Qc1AaD1AZh1nt5PP7fjeRs2k1SQhxtEwIkJQRISoir+15vO/7z120TArSpe52cGE92p2RPRvrhjNDHALnOuTwAM5sDTAbqB/pk4CkXWoTwIzNrb2bdnHNFEa9YRKQRurdvS/f2bZvURq1zvLe+hK17DlBZXcuB6loqqmupqA5SUVNLuMuw/viC45h6et8m1dKQcAI9C8ivt13Al0ffDR2TBXwh0M1sGjANoFevXsdaq4hIVN123kBuO29gg/ucc1TWBKmsC/eDQf956Ie2H3h1LRlpbTypL5xAb+j/Cw79dyicY3DOzQRmAowaNUpLiouIb5jZZ1Mu6SQc9riJQ7t6VkM4T9QpAHrW2+4BFDbiGBER8VA4gZ4DDDCzPmaWCFwGzDvkmHnANRYyFijV/LmISPM66pSLc67GzG4GFhC6bPFx59wqM5tet38GMJ/QJYu5hC5bvN67kkVEpCFhXYfunJtPKLTrvzej3msHfCeypYmIyLHQU+lFRHxCgS4i4hMKdBERn1Cgi4j4hLlw71WNdMdmJcDmqHQeWzoDO6JdRAzR+ficzsUX6XyE9HbOZTS0I2qBLiFmttg5NyradcQKnY/P6Vx8kc7H0WnKRUTEJxToIiI+oUCPvpnRLiDG6Hx8Tufii3Q+jkJz6CIiPqERuoiITyjQRUR8QoHejMysp5m9ZWZrzGyVmd1S935HM3vdzNbXfe8Q7Vqbi5kFzOwTM3u5brs1n4v2Zva8ma2t+x05pbWeDzP7ft3fyEoz+5uZJbXWc3EsFOjNqwa43Tl3HDAW+I6ZDQHuAt5wzg0A3qjbbi1uAdbU227N5+Ih4FXn3GBgOKHz0urOh5llAd8DRjnnhhJ6bPdltMJzcawU6M3IOVfknPu47nUZoT/YLEKLbM+uO2w28LWoFNjMzKwHcCEwq97brfVctANOBx4DcM5VOef20ErPB6FHe7c1s3ggmdAKaK31XIRNgR4lZpYNnAgsBLocXOGp7ntmFEtrTn8A7gSC9d5rreeiL1ACPFE3BTXLzFJohefDObcVeBDYQmih+VLn3Gu0wnNxrBToUWBmqcALwK3Oub3RricazOwrQLFzbkm0a4kR8cBJwKPOuROBfbTSKYW6ufHJQB+gO5BiZldFt6qWQYHezMwsgVCYP+uce7Hu7e1m1q1ufzegOFr1NaPTgK+a2SZgDnC2mT1D6zwXEFpovcA5t7Bu+3lCAd8az8e5wEbnXIlzrhp4ETiV1nkujokCvRmZmRGaI13jnPtdvV3zgGvrXl8LzG3u2pqbc+5u51wP51w2oQ+83nTOXUUrPBcAzrltQL6ZDap76xxgNa3zfGwBxppZct3fzDmEPm9qjefimOhO0WZkZuOA94AVfD5v/CNC8+jPAb0I/TJf6pzbFZUio8DMzgR+4Jz7ipl1opWeCzMbQegD4kQgj9Bi63G0wvNhZj8DvknoyrBPgClAKq3wXBwLBbqIiE9oykVExCcU6CIiPqFAFxHxCQW6iIhPKNBFRHxCgS4i4hMKdBERn/j/E9rnERYHZFwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "x=a[\"Group Size\"]\n",
    "y=a[\"Indel Thresh\"]\n",
    "\n",
    "plt.plot(x,y)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-17T23:31:04.485332Z",
     "start_time": "2022-06-17T23:31:04.477218Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2861.02294921875"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "3e9/4**10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-17T23:32:23.398400Z",
     "start_time": "2022-06-17T23:32:23.382128Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.027938280254602422\n",
      "15.839999947373023\n",
      "3519999988.305116\n"
     ]
    }
   ],
   "source": [
    "def estimate_perc_full(size,key_length, ex_length):\n",
    "    \n",
    "    samps = size-(key_length*size/ex_length)\n",
    "    expected_unique=(1-(1-1/samps)**(4**key_length))*samps\n",
    "    return expected_unique/4**key_length\n",
    "\n",
    "def estimate_mem_req(exome_mem, size, key_length, ex_length):\n",
    "    \n",
    "    perc_full=estimate_perc_full(size, key_length, ex_length)\n",
    "    \n",
    "    if perc_full>1:\n",
    "        return (4**key_length)/size*exome_mem\n",
    "        \n",
    "    return perc_full*exome_mem*key_length*(4**key_length/size)\n",
    "\n",
    "a\n",
    "print(estimate_perc_full(3e7,15,3e5))\n",
    "print(estimate_mem_req(1,4e9,18,150))\n",
    "print(estimate_perc_full(4e9,18,150)*4**18)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-29T23:44:59.898429Z",
     "start_time": "2022-06-29T23:44:59.875153Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1048576"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "4**10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-29T23:17:21.304040Z",
     "start_time": "2022-06-29T23:17:21.292772Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.72830220637843e-05\n"
     ]
    }
   ],
   "source": [
    "def estimate_perc_full_2(bases,key_length, num_ex,sparsity_factor):\n",
    "    \n",
    "    \n",
    "    samps = ((bases/num_ex)-key_length)*num_ex/sparsity_factor\n",
    "    expected_unique=(1-(1-1/samps)**(4**key_length))*samps\n",
    "    return expected_unique/4**key_length\n",
    "\n",
    "def estimate_perc_full_3(bases,key_length, num_ex,sparsity_factor):\n",
    "    \n",
    "    \n",
    "    samps = ((bases/num_ex)-key_length)*num_ex/sparsity_factor\n",
    "    expected_unique=(1-(1-1/samps)**(4**key_length))*samps\n",
    "    return expected_unique/4**key_length\n",
    "\n",
    "print(estimate_perc_full_2(3e7,20,3e7/3e5,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-17T01:23:12.268457Z",
     "start_time": "2022-06-17T01:23:12.262474Z"
    }
   },
   "outputs": [],
   "source": [
    "def choose_genome_sparsity(genome_filesize, read_filesize, read_length, key_size, average_exon_size, flex_factor, num_files):\n",
    "    \"\"\"\n",
    "    function to choose genome indices and read indices to optimize for runtime\n",
    "    genome and read filesizes only need to be in the same units\n",
    "    \"\"\"\n",
    "    \n",
    "    if key_size>=average_exon_size or key_size >= read_length:\n",
    "        return(\"err: key size is too large\")\n",
    "    \n",
    "    num = genome_filesize*(1-(key_size/average_exon_size))\n",
    "    denom = (read_filesize/read_length)*flex_factor*num_files\n",
    "    \n",
    "    return int((num/denom)**0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-17T01:25:17.640293Z",
     "start_time": "2022-06-17T01:25:17.626334Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "choose_genome_sparsity(3,9,100,30,100,2,1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-06-01T21:43:55.458865Z",
     "start_time": "2022-06-01T21:43:55.254211Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21844 5461\n"
     ]
    }
   ],
   "source": [
    "a=[]\n",
    "seen_key = set()\n",
    "seen_slot = set()\n",
    "\n",
    "for i in range(10000):\n",
    "    \n",
    "    key=(str(np.random.randint(1,5)*1000000+np.random.randint(1,5)*100000+np.random.randint(1,5)*10000+np.random.randint(1,5)*1000+np.random.randint(1,5)*100+np.random.randint(1,5)*10+np.random.randint(1,5)))\n",
    "    \n",
    "    if key not in seen_key:\n",
    "        seen_key.add(key)\n",
    "        \n",
    "        slot=(int(key[0]))+int(key[1])*4+int(key[2])*16+int(key[3])*64+int(key[4])*256+int(key[5])*1024+int(key[6])*4096\n",
    "        if slot in seen_slot:\n",
    "            print(f\"bummer: {key,slot}\")\n",
    "            \n",
    "        else:\n",
    "            seen_slot.add(slot)\n",
    "\n",
    "            \n",
    "print(max(seen_slot), min(seen_slot))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-05-30T21:04:44.117Z"
    }
   },
   "outputs": [],
   "source": [
    "a=[]\n",
    "seen_key = set()\n",
    "seen_slot = set()\n",
    "\n",
    "for i in range(10000):\n",
    "    \n",
    "    key=(str(np.random.randint(0,4)*100+np.random.randint(0,4)*10+np.random.randint(0,4)))\n",
    "    \n",
    "    if key not in seen_key:\n",
    "        seen_key.add(key)\n",
    "        \n",
    "        slot=(int(key[0])*4)+int(key[1])*16+int(key[2])+10\n",
    "        if slot in seen_slot:\n",
    "            print(f\"bummer: {key,slot}\")\n",
    "            \n",
    "        else:\n",
    "            seen_slot.add(slot)\n",
    "            \n",
    "print(\"done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2022-08-09T06:37:45.274226Z",
     "start_time": "2022-08-09T06:37:45.262227Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'12345678'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a='12345678910'\n",
    "\n",
    "a[:8]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": true
   },
   "source": [
    "# Old Implementation: read arrays rather than strings, seems 5x slower"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "ExecuteTime": {
     "start_time": "2022-05-30T21:04:44.119Z"
    },
    "hidden": true
   },
   "outputs": [],
   "source": [
    "def create_genome_exons(total_size, exon_size):\n",
    "    \"\"\"\n",
    "    create syntheic exon to map to\n",
    "    \n",
    "    input:total_size - length of genome\n",
    "    exon_size- length of each exon\n",
    "    output: matrix of exons\n",
    "    \"\"\"\n",
    "    \n",
    "    exome = np.random.randint(1,5,total_size)\n",
    "    exome = exome.reshape(-1,exon_size)\n",
    "    \n",
    "    return exome\n",
    "\n",
    "def create_transcriptome(read_length, reads):\n",
    "    \"\"\"\n",
    "    create synthetic transcriptome\n",
    "    input-read length\n",
    "    input- reads\n",
    "    \n",
    "    output- reads number of random transcripts of read_length\n",
    "    \"\"\"\n",
    "    \n",
    "    transcriptome = np.random.randint(1,5,read_length*reads)\n",
    "    transcriptome = transcriptome.reshape(-1,reads)\n",
    "    \n",
    "    return transcriptome\n",
    "\n",
    "def create_matched_transcriptome(exome, read_length):\n",
    "    \"\"\"\n",
    "    create synthetic matched reads\n",
    "    \n",
    "    input: exome (mxn)\n",
    "    input: read_length: int\n",
    "    \"\"\"\n",
    "    \n",
    "    exome=exome.flatten()\n",
    "    out_matrix = np.zeros([int(len(exome)-read_length)+1, read_length])\n",
    "    \n",
    "    for i in range(len(exome)-read_length+1):\n",
    "        transcript = exome[i:i+read_length]\n",
    "        out_matrix[i]+=transcript\n",
    "        \n",
    "    return out_matrix\n",
    "\n",
    "def create_indels(matched_transcriptome, indels_per_read):\n",
    "    \n",
    "    outmatrix = np.zeros([matched_transcriptome.shape[0],matched_transcriptome.shape[1]])\n",
    "    \n",
    "    for row in range(matched_transcriptome.shape[0]):\n",
    "        \n",
    "        new_read = matched_transcriptome[row]\n",
    "        \n",
    "        change_index = np.random.randint(0,len(new_read))\n",
    "        \n",
    "        if new_read[change_index]==1:\n",
    "            new_read[change_index]=4\n",
    "            \n",
    "        else:\n",
    "            new_read[change_index]-=1\n",
    "            \n",
    "        outmatrix[row]+=new_read\n",
    "        \n",
    "    return outmatrix\n",
    "\n",
    "        \n",
    "class read_object():\n",
    "    \"\"\"\n",
    "    create object to store the read we generate and its type...may expand class in the future to incorporate all\n",
    "    the read types we need as well as method to iterate over them.\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__(self, variety, exome, read_length, indels_per_read = 1):\n",
    "        \n",
    "        self.group = variety\n",
    "        self.indels_per_read = indels_per_read\n",
    "        \n",
    "        #create indel read\n",
    "        if self.group ==\"indel\":\n",
    "            self.reads = create_indels(create_matched_transcriptome(exome, read_length),self.indels_per_read)\n",
    "        \n",
    "        #create good transcriptome read   \n",
    "        elif self.group ==\"good\":\n",
    "            self.reads = create_matched_transcriptome(exome, read_length)\n",
    "        \n",
    "        #create bad transcriptome read\n",
    "        elif self.group == \"bad\":\n",
    "            self.reads = create_transcriptome (exome.shape[0]*read_length, read_length)\n",
    "            \n",
    "class filter_object():\n",
    "    \"\"\"\n",
    "    object to house the filter type and its attributes\n",
    "    \"\"\"\n",
    "    \n",
    "    def __init__ (self, exome, group_size, fwd_move, spacing=0):\n",
    "        self.exome = exome\n",
    "        self.group_size = group_size\n",
    "        self.fwd_move = fwd_move\n",
    "        self.spacing = spacing\n",
    "    \n",
    "    def generate_filter(self, verbose = False):\n",
    "        \"\"\"\n",
    "        inputs:\n",
    "        exome: exome matrix form\n",
    "        group_size: number of bases in partial key to populate filter with\n",
    "        spacing: number of bases to skip when generating key\n",
    "        fwd_move: how many bp to move forward following partial key creation\n",
    "        filter object: filter array to be populated with seen reads, can experient with augmenting existing filters later\n",
    "        could have if/then stacked filter where we just keep grabbing larger partial keys (might be more efficient way\n",
    "        to access reads and score them)\n",
    "\n",
    "        output: filter with corresponding cells set to 1 for reads seen\n",
    "        \"\"\"\n",
    "\n",
    "        #generate proper size for array (can do this more compactly)\n",
    "        base=0\n",
    "        for i in range(self.group_size):\n",
    "            base+=(4*10**i)\n",
    "\n",
    "        if verbose:\n",
    "            print(\"base size: \"+str(base))\n",
    "\n",
    "\n",
    "        #generate filter\n",
    "        filter_array = np.zeros(base+1)\n",
    "\n",
    "        #grab each exon and create keys to populate filter\n",
    "        for row in range(self.exome.shape[0]):\n",
    "\n",
    "            exon = self.exome[row]\n",
    "\n",
    "            if verbose:\n",
    "                print(f'len exon: {len(exon)}')\n",
    "\n",
    "            if self.spacing >0:\n",
    "                #set this aside for now\n",
    "                pass\n",
    "\n",
    "            else:\n",
    "                start_key=0\n",
    "\n",
    "                #grab partial keys until we reach the end of the exon\n",
    "                while (start_key+self.group_size)<=len(exon):\n",
    "\n",
    "                    if verbose:\n",
    "                        print(f'key: {exon[start_key:start_key+self.group_size]}')\n",
    "\n",
    "                    #generate key\n",
    "                    key=0\n",
    "                    mult=1\n",
    "                    for i in range(self.group_size):\n",
    "                        key+=exon[start_key+i]*mult\n",
    "                        mult*=10\n",
    "                      \n",
    "                    #add key to filter array\n",
    "                    filter_array[key]=1\n",
    "\n",
    "                    #move the start of next partial key according to fwd move input\n",
    "                    start_key+=self.fwd_move\n",
    "\n",
    "                    if verbose:\n",
    "\n",
    "                        print(f'new start key{start_key}')\n",
    "\n",
    "        self.filter_ = filter_array\n",
    "        \n",
    "        \n",
    "    def check_read(self, read, check_indices, verbose = False):\n",
    "        \"\"\"\n",
    "        function checks to see if a read is present in the filter, utilizing self.filter_, group size and fwd_move\n",
    "        \n",
    "        read: transcriptome read to be mapped\n",
    "        check_indices: array of start positions that we want to check\n",
    "        \n",
    "        outputs:number of size(group size) partial keys that appear in our filter\n",
    "        \"\"\"\n",
    "        #test all combos if check indices are not stipulated\n",
    "        if check_indices == False:\n",
    "            check_indices = range(len(read)-self.group_size+1)\n",
    "            \n",
    "        #int to store number of partial reads that map\n",
    "        hits = 0\n",
    "        \n",
    "        #catch this later\n",
    "        if self.spacing >0:\n",
    "            pass\n",
    "        \n",
    "        #grab partial keys for all given start keys\n",
    "        for start_key in check_indices:\n",
    "\n",
    "            if verbose:\n",
    "                print(f'key: {read[start_key:start_key+self.group_size]}')\n",
    "\n",
    "            #generate key\n",
    "            key=0\n",
    "            mult=1\n",
    "            for i in range(self.group_size):\n",
    "                   \n",
    "                key+=read[start_key+i]*mult\n",
    "                mult*=10\n",
    "\n",
    "            #check filter_ for key\n",
    "            hits += self.filter_[int(key)]   \n",
    "\n",
    "            #move the start of next partial key according to fwd move input\n",
    "            start_key+=self.fwd_move\n",
    "\n",
    "            if verbose:\n",
    "                print(f'new start key{start_key}')\n",
    "                    \n",
    "        return hits\n",
    "\n",
    "       \n",
    "def test_filters(read_groups, exome, group_size_array, read_indices_array, fwd_move_array, thresh_array, spacing=0):\n",
    "    \"\"\"\n",
    "    function to test filter params, going to begin by varying key size and forward moves\n",
    "    \n",
    "    inputs:\n",
    "    reads: array of references to arrays holding reads\n",
    "    exome: exome to create filter with \n",
    "    group_size_array: array holding group sizes to try\n",
    "    fwd_move_array: array holding fwd moves to try\n",
    "    perc_thresh: percent threshold above which we want to count reads\n",
    "    read_indices_array: read_indices to try...need to figure out how to pass rules thru array to subsample\n",
    "    spacing: set to 0 for now\n",
    "    \n",
    "    output: dataframe with summary statistics for accuracy and time\n",
    "    \"\"\"\n",
    "    \n",
    "    #initialize output array variables\n",
    "    gs = []\n",
    "    fwd = []\n",
    "    ri = []\n",
    "    good_percentage = []\n",
    "    good_variance = []\n",
    "    bad_percentage = []\n",
    "    bad_variance = []\n",
    "    good_thresh = []\n",
    "    bad_thresh = []\n",
    "    indel_thresh = []\n",
    "    indel_percentage = []\n",
    "    indel_variance = []\n",
    "    generate_time_array=[]\n",
    "    check_time_array=[]\n",
    "    exome_shape =[]\n",
    "    filter_full = []\n",
    "    read_lengths = []\n",
    "    num_reads=[]\n",
    "    \n",
    "    #loop over grid\n",
    "    for group_size in group_size_array:\n",
    "        for fwd_move in fwd_move_array:\n",
    "            for read_indices in read_indices_array:\n",
    "                for perc_thresh in thresh_array:\n",
    "                    gs.append(group_size)\n",
    "                    fwd.append(fwd_move)\n",
    "                    ri.append(read_indices)\n",
    "\n",
    "                    #create filter\n",
    "                    filter_obj = filter_object(exome, group_size, fwd_move)\n",
    "\n",
    "\n",
    "                    #time how long it takes to create \n",
    "                    generate_beg = time.time()\n",
    "                    filter_obj.generate_filter()\n",
    "                    generate_end = time.time()\n",
    "\n",
    "                    #see how full the filter is\n",
    "                    filter_full_perc= sum(filter_obj.filter_)/4**group_size\n",
    "\n",
    "\n",
    "\n",
    "                    #append create time to output array, grabbing percentage of array that is filled\n",
    "                    generate_time_array.append(generate_end-generate_beg)\n",
    "\n",
    "                    #TODO generate read indices according to protocol\n",
    "\n",
    "                    #test all the transcriptome read objects and time them\n",
    "                    check_beg = time.time()\n",
    "                    for read_group in read_groups:\n",
    "\n",
    "                        #creat placeholder variable to store read by read results\n",
    "                        out_array = []\n",
    "                        for row in range(read_group.reads.shape[0]):\n",
    "\n",
    "                            #grab individual read and check it in the filter\n",
    "                            read = read_group.reads[row]\n",
    "                            out_array.append(filter_obj.check_read(read, read_indices))\n",
    "\n",
    "                            if out_array[-1]<1:\n",
    "                                print(f\"hey!-----{row,read}\")\n",
    "\n",
    "                        #establish number of partial keys we tried\n",
    "                        if read_indices == False:\n",
    "                            pos = len(read_group.reads[0])-group_size+1\n",
    "\n",
    "                        else:\n",
    "                            pos = len(read_indices)\n",
    "\n",
    "                        #calculate ave percentage of hits\n",
    "                        percentage_array = [i/pos for i in out_array]\n",
    "                        prec = sum(percentage_array)/len(percentage_array)\n",
    "\n",
    "                        #calcualte percentage of reads above thresh\n",
    "                        thresh=0 \n",
    "                        for i in percentage_array:\n",
    "                            if i >=perc_thresh:\n",
    "                                thresh+=1\n",
    "\n",
    "                        thresh = thresh /len(percentage_array)\n",
    "\n",
    "\n",
    "                        #calculate variance of num hits\n",
    "                        var = np.var(out_array)\n",
    "\n",
    "\n",
    "                        if read_group.group == \"indel\":\n",
    "                            indel_percentage.append(prec)\n",
    "                            indel_variance.append(var)\n",
    "                            indel_thresh.append(thresh)\n",
    "\n",
    "                        elif read_group.group == \"good\":\n",
    "                            good_percentage.append(prec)\n",
    "                            good_variance.append(var)\n",
    "                            good_thresh.append(thresh)\n",
    "\n",
    "                        else:\n",
    "                            bad_percentage.append(prec)\n",
    "                            bad_variance.append(var)\n",
    "                            bad_thresh.append(thresh)\n",
    "\n",
    "                    check_end = time.time()\n",
    "                    check_time_array.append(check_end-check_beg)\n",
    "                    exome_shape.append(exome.shape[0])\n",
    "                    filter_full.append(filter_full_perc)\n",
    "                    read_lengths.append(read_groups[0].reads.shape[1])\n",
    "                    num_reads.append(read_groups[0].reads.shape[0])\n",
    "\n",
    "\n",
    "        out_df = pd.DataFrame({\"Group Size\":gs,\"Fwd Move\":fwd,\"Read Indices\":ri, \"Good Percentage\":good_percentage,\n",
    "                              \"Bad Percentage\":bad_percentage, \"Indel Percentage\":indel_percentage, \"Good Variance\":good_variance,\n",
    "                              \"Bad Variance\":bad_variance, \"Indel Variance\":indel_variance, \"Good Thresh\":good_thresh,\n",
    "                              \"Bad Thresh\":bad_thresh, \"Indel Thresh\": indel_thresh, \"Create Time\": generate_time_array,\n",
    "                              \"Check Time\": check_time_array,\"Exome Size\":exome_shape, \"Read Lengths\": read_lengths, \"Num Reads\":num_reads, \n",
    "                              \"Filter Full Percentage\":filter_full})\n",
    "        \n",
    "    return out_df, filter_obj      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "hidden": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
